{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/VJalili/temp_test_notebook/blob/test/PZM_Parse_VCF_files.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Setup gsutil and AuthNZ"
      ],
      "metadata": {
        "id": "8BiH6C-QOVyk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# !gcloud init"
      ],
      "metadata": {
        "id": "91_ncxBgGpa4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from subprocess import check_output\n",
        "def object_exists(bucket):\n",
        "  out = check_output(f\"gsutil -q stat {bucket}; echo $?\", shell=True).decode(\"utf-8\").strip()\n",
        "  return True if out == \"0\" else False\n",
        "\n",
        "def move_object(source, dest):\n",
        "  out = check_output(f\"gsutil -m mv {source} {dest}\", shell=True).decode(\"utf-8\").strip()\n",
        "  return out\n",
        "\n",
        "def copy_object(source, dest):\n",
        "  out = check_output(f\"gsutil -m cp {source} {dest}\", shell=True).decode(\"utf-8\").strip()\n",
        "  return out"
      ],
      "metadata": {
        "id": "TPvodZbiaHTP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### TEST BLOCK\n",
        "this is test"
      ],
      "metadata": {
        "id": "nkcYWe-mHh4Y"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install test\n",
        "import test_package"
      ],
      "metadata": {
        "id": "E2ZMpih1Hkcx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Setup Google Storage Client"
      ],
      "metadata": {
        "id": "EepirIeRHUch"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# !gcloud auth application-default login"
      ],
      "metadata": {
        "id": "GylvCcOjHbCq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from google.cloud import storage\n",
        "storage_client = storage.Client(project=\"broad-dsde-methods\")\n",
        "\n",
        "# This method uses the Google cloud storage library which performs faster\n",
        "# calling gsutil using `check_output`.\n",
        "def copy_object_to_bucket(bucket_name, blob_name, source_filename):\n",
        "  bucket = storage_client.get_bucket(bucket_name)\n",
        "  blob = bucket.blob(blob_name)\n",
        "  blob.upload_from_filename(source_filename)"
      ],
      "metadata": {
        "id": "oXdDLHgqGt3x"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# !pip install pysam"
      ],
      "metadata": {
        "id": "aFNJgiB5DQHO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "j_nEe_8ccP0M"
      },
      "source": [
        "### Import python packages"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "1mFSYuORcSRn"
      },
      "outputs": [],
      "source": [
        "import subprocess\n",
        "import gspread\n",
        "import copy\n",
        "from oauth2client.client import GoogleCredentials\n",
        "import json\n",
        "import pandas as pd\n",
        "from pathlib import Path\n",
        "from dataclasses import dataclass, field\n",
        "from datetime import datetime\n",
        "import textwrap\n",
        "from matplotlib.pyplot import figure\n",
        "import gzip\n",
        "from dataclasses import dataclass\n",
        "from typing import Callable, Any, Dict, Tuple\n",
        "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "from matplotlib import pyplot\n",
        "import os\n",
        "from os import listdir\n",
        "import os.path\n",
        "from os.path import isfile, join\n",
        "from pathlib import Path\n",
        "from numpy.lib.function_base import average\n",
        "import pysam\n",
        "from pysam import VariantFile\n",
        "import tensorflow as tf\n",
        "import base64\n",
        "from urllib.parse import urljoin\n",
        "from sklearn.cluster import KMeans\n",
        "from sklearn.manifold import TSNE\n",
        "from typing import Callable\n",
        "import matplotlib.pyplot as plt\n",
        "from matplotlib import collections\n",
        "from sklearn.metrics import confusion_matrix\n",
        "\n",
        "from dataclasses import dataclass\n",
        "\n",
        "from tqdm import tqdm\n",
        "import tqdm.notebook as tq\n",
        "\n",
        "# import upsetplot\n",
        "# from upsetplot import plot\n",
        "import math\n",
        "\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "custom_style = {\"axes.facecolor\": \"#EAEAF2\"}\n",
        "sns.set_style(rc=custom_style)\n",
        "sns.color_palette(\"Set2\")\n",
        "\n",
        "plt.rcParams[\"axes.facecolor\"]=\"#ffffff\"\n",
        "plt.rcParams[\"savefig.facecolor\"]=\"#ffffff\""
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "run_test = True #@param {type:\"boolean\"}"
      ],
      "metadata": {
        "id": "hP5DqFrWGmXZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Helper methods"
      ],
      "metadata": {
        "id": "QyNOnxLeLiLR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "working_dir = os.path.join(\".\", \"pzm\")"
      ],
      "metadata": {
        "id": "whdUmtxBdf8M"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Get and read the data table.\n",
        "gs_bucket = \"gs://broad-dsde-methods-vj/pzm/\"\n",
        "data_table_filename = os.path.join(working_dir, \"data_table.tsv\")\n",
        "if not isfile(data_table_filename):\n",
        "  copy_object(gs_bucket + \"data_table.tsv\", data_table_filename)\n",
        "\n",
        "data_table = pd.read_csv(data_table_filename, sep=\"\\t\")\n",
        "\n",
        "sample_id_h = \"sample_id\" # sample id column header\n",
        "data_table = data_table.rename(columns={\"entity:sample_id\": sample_id_h})"
      ],
      "metadata": {
        "id": "B4_d5WKTdZpV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "data_table.shape"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "aPe_inELeTD2",
        "outputId": "16903b75-b21b-42a5-d0f0-6c6f9560bd63"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(1951, 61)"
            ]
          },
          "metadata": {},
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "data_table.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 473
        },
        "id": "VxFWC0LOjEMq",
        "outputId": "237f7b8b-8157-4c18-c2fd-689ece7acf29"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "   sample_id                                           RM_sites  \\\n",
              "0  SS0012986  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "1  SS0012989  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "2  SS0012999  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "3  SS0013012  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "4  SS0013013  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "\n",
              "                                          RM_summary    WES_ID  \\\n",
              "0  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...       NaN   \n",
              "1  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...       NaN   \n",
              "2  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...  12163.p1   \n",
              "3  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...       NaN   \n",
              "4  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...  13332.p1   \n",
              "\n",
              "                                        WES_call_idx  \\\n",
              "0  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "1  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "2  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "3  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "4  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "\n",
              "                                        WES_call_vcf  \\\n",
              "0  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "1  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "2  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "3  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "4  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "\n",
              "                                        WES_eval_idx  \\\n",
              "0  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "1  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "2  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "3  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "4  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...   \n",
              "\n",
              "                                        WES_eval_vcf analyte_type  \\\n",
              "0  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...          DNA   \n",
              "1  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...          DNA   \n",
              "2  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...          DNA   \n",
              "3  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...          DNA   \n",
              "4  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...          DNA   \n",
              "\n",
              "                                 anno_rm_gc_pass_vcf  ...  \\\n",
              "0  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...  ...   \n",
              "1  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...  ...   \n",
              "2  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...  ...   \n",
              "3  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...  ...   \n",
              "4  gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...  ...   \n",
              "\n",
              "                                 original_cram_index participant  \\\n",
              "0  gs://fc-secure-22693da1-7a53-4115-9b93-98856c7...    13867.s2   \n",
              "1  gs://fc-secure-22693da1-7a53-4115-9b93-98856c7...         NaN   \n",
              "2  gs://fc-secure-22693da1-7a53-4115-9b93-98856c7...         NaN   \n",
              "3  gs://fc-secure-22693da1-7a53-4115-9b93-98856c7...    11420.s1   \n",
              "4  gs://fc-secure-22693da1-7a53-4115-9b93-98856c7...         NaN   \n",
              "\n",
              "  participant_id  read_orientation_model_params  relation  sequencing_assay  \\\n",
              "0       13867.s2                            NaN   sibling               WGS   \n",
              "1       13521.s1                            NaN   sibling               WGS   \n",
              "2       12163.p1                            NaN   proband               WGS   \n",
              "3       11420.s1                            NaN   sibling               WGS   \n",
              "4       13332.p1                            NaN   proband               WGS   \n",
              "\n",
              "          sequencing_id     sex  submitter_id wes_id  \n",
              "0  SS0012986.final.cram    Male          NYGC      .  \n",
              "1  SS0012989.final.cram  Female          NYGC    NaN  \n",
              "2  SS0012999.final.cram    Male          NYGC    NaN  \n",
              "3  SS0013012.final.cram  Female          NYGC    NaN  \n",
              "4  SS0013013.final.cram    Male          NYGC    NaN  \n",
              "\n",
              "[5 rows x 61 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-55021a31-f8fb-49ae-9402-02e95fdfe907\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sample_id</th>\n",
              "      <th>RM_sites</th>\n",
              "      <th>RM_summary</th>\n",
              "      <th>WES_ID</th>\n",
              "      <th>WES_call_idx</th>\n",
              "      <th>WES_call_vcf</th>\n",
              "      <th>WES_eval_idx</th>\n",
              "      <th>WES_eval_vcf</th>\n",
              "      <th>analyte_type</th>\n",
              "      <th>anno_rm_gc_pass_vcf</th>\n",
              "      <th>...</th>\n",
              "      <th>original_cram_index</th>\n",
              "      <th>participant</th>\n",
              "      <th>participant_id</th>\n",
              "      <th>read_orientation_model_params</th>\n",
              "      <th>relation</th>\n",
              "      <th>sequencing_assay</th>\n",
              "      <th>sequencing_id</th>\n",
              "      <th>sex</th>\n",
              "      <th>submitter_id</th>\n",
              "      <th>wes_id</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SS0012986</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>DNA</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>...</td>\n",
              "      <td>gs://fc-secure-22693da1-7a53-4115-9b93-98856c7...</td>\n",
              "      <td>13867.s2</td>\n",
              "      <td>13867.s2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sibling</td>\n",
              "      <td>WGS</td>\n",
              "      <td>SS0012986.final.cram</td>\n",
              "      <td>Male</td>\n",
              "      <td>NYGC</td>\n",
              "      <td>.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SS0012989</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>DNA</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>...</td>\n",
              "      <td>gs://fc-secure-22693da1-7a53-4115-9b93-98856c7...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>13521.s1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sibling</td>\n",
              "      <td>WGS</td>\n",
              "      <td>SS0012989.final.cram</td>\n",
              "      <td>Female</td>\n",
              "      <td>NYGC</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SS0012999</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>12163.p1</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>DNA</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>...</td>\n",
              "      <td>gs://fc-secure-22693da1-7a53-4115-9b93-98856c7...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>12163.p1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>proband</td>\n",
              "      <td>WGS</td>\n",
              "      <td>SS0012999.final.cram</td>\n",
              "      <td>Male</td>\n",
              "      <td>NYGC</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SS0013012</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>DNA</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>...</td>\n",
              "      <td>gs://fc-secure-22693da1-7a53-4115-9b93-98856c7...</td>\n",
              "      <td>11420.s1</td>\n",
              "      <td>11420.s1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sibling</td>\n",
              "      <td>WGS</td>\n",
              "      <td>SS0013012.final.cram</td>\n",
              "      <td>Female</td>\n",
              "      <td>NYGC</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SS0013013</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>13332.p1</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>DNA</td>\n",
              "      <td>gs://fc-0e8b9bf4-7749-4228-9bd5-4cdb2b81d3b3/s...</td>\n",
              "      <td>...</td>\n",
              "      <td>gs://fc-secure-22693da1-7a53-4115-9b93-98856c7...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>13332.p1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>proband</td>\n",
              "      <td>WGS</td>\n",
              "      <td>SS0013013.final.cram</td>\n",
              "      <td>Male</td>\n",
              "      <td>NYGC</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 61 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-55021a31-f8fb-49ae-9402-02e95fdfe907')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  \n",
              "    <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-55021a31-f8fb-49ae-9402-02e95fdfe907')\"\n",
              "            title=\"Generate charts.\"\n",
              "            style=\"display:none;\">\n",
              "      \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "      <g>\n",
              "          <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "      </g>\n",
              "  </svg>\n",
              "    </button>\n",
              "   \n",
              "  <style>\n",
              "    .colab-df-quickchart {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-quickchart:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-quickchart {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-quickchart:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#df-55021a31-f8fb-49ae-9402-02e95fdfe907 button.colab-df-quickchart');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function quickchart(key) {\n",
              "        const containerElement = document.querySelector('#df-55021a31-f8fb-49ae-9402-02e95fdfe907');\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'generateCharts', [key], {})\n",
              "      }\n",
              "    </script>    \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-55021a31-f8fb-49ae-9402-02e95fdfe907 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-55021a31-f8fb-49ae-9402-02e95fdfe907');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "data_table.columns"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "QTTH7iMRe4N9",
        "outputId": "3d5f98b2-3041-4770-d684-c668e1c73128"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Index(['sample_id', 'RM_sites', 'RM_summary', 'WES_ID', 'WES_call_idx',\n",
              "       'WES_call_vcf', 'WES_eval_idx', 'WES_eval_vcf', 'analyte_type',\n",
              "       'anno_rm_gc_pass_vcf', 'anno_rm_gc_pass_vcf_idx', 'anno_rm_gc_vcf',\n",
              "       'anno_rm_gc_vcf_idx', 'bamout', 'bamout_i', 'bamout_index',\n",
              "       'confirmed_wes_wgs_rel', 'contamination_table', 'family_id',\n",
              "       'father_id', 'filtered_vcf', 'filtered_vcf_idx', 'filtered_zip_vcf',\n",
              "       'filtered_zip_vcf_idx', 'filtering_stats', 'full_anno_pass_vcf',\n",
              "       'full_anno_pass_vcf_idx', 'full_anno_vcf', 'full_anno_vcf_idx',\n",
              "       'funcotated_file', 'funcotated_file_index', 'maf_segments',\n",
              "       'mod_filtered_vcf', 'mod_filtered_vcf_idx', 'mother_id',\n",
              "       'mutect_WES_call', 'mutect_WES_eval', 'mutect_sites',\n",
              "       'mutect_sites_no_SR_SD', 'mutect_sites_no_repeat', 'mutect_stats',\n",
              "       'mutect_summary', 'mutect_summary_WES_call', 'mutect_summary_WES_eval',\n",
              "       'mutect_summary_no_SR_SD', 'mutect_summary_no_repeat', 'no_SR_SD_vcf',\n",
              "       'no_SR_SD_vcf_idx', 'no_repeat_vcf', 'no_repeat_vcf_idx',\n",
              "       'original_cram', 'original_cram_index', 'participant', 'participant_id',\n",
              "       'read_orientation_model_params', 'relation', 'sequencing_assay',\n",
              "       'sequencing_id', 'sex', 'submitter_id', 'wes_id'],\n",
              "      dtype='object')"
            ]
          },
          "metadata": {},
          "execution_count": 12
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def localize_obj_if_needed(sample_id):\n",
        "  local_filename = os.path.join(working_dir, f\"{sample_id}.vcf.gz\")\n",
        "  if not isfile(local_filename):\n",
        "    copy_object(data_table_dict[sample_id], local_filename)\n",
        "\n",
        "  unfiltered_local_filename = os.path.join(working_dir, f\"unfiltered_{sample_id}.vcf.gz\")\n",
        "  if not isfile(unfiltered_local_filename):\n",
        "    copy_object(unfiltered_data_table_dict[sample_id], unfiltered_local_filename)\n",
        "\n",
        "  unfiltered_idx_local_filename = os.path.join(working_dir, f\"unfiltered_{sample_id}.vcf.gz.tbi\")\n",
        "  if not isfile(unfiltered_idx_local_filename):\n",
        "    copy_object(unfiltered_idx_table_dict[sample_id], unfiltered_idx_local_filename)\n",
        "\n",
        "  return local_filename, unfiltered_local_filename"
      ],
      "metadata": {
        "id": "mbZelYk9VO_F"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "vcf_column = \"full_anno_pass_vcf\" # \"anno_rm_gc_pass_vcf\"\n",
        "unfiltered_vcf_column = \"full_anno_vcf\" # \"anno_rm_gc_vcf\"\n",
        "data_table_dict = dict(zip(data_table[sample_id_h], data_table[vcf_column]))\n",
        "unfiltered_data_table_dict = dict(zip(data_table[sample_id_h], data_table[unfiltered_vcf_column]))\n",
        "unfiltered_idx_table_dict = dict(zip(data_table[sample_id_h], data_table[unfiltered_vcf_column + \"_idx\"]))\n",
        "# sample_ids = [\"SS0012986\", \"SS0012989\", \"SSC09045\", \"SSC05448\", \"SSC02667\", \"SSC06504\", \"SSC09053\", \"SSC04423\"]\n",
        "sample_ids = list(data_table_dict.keys())\n",
        "# samples = {}\n",
        "# unfiltered_samples = {}\n",
        "\n",
        "# for sample_id in sample_ids:\n",
        "#   local_filename, local_unfiltered_filename = localize_obj_if_needed(sample_id)\n",
        "#   sample_ids[sample_id] = local_filename\n",
        "#   unfiltered_samples[sample_id] = local_unfiltered_filename\n",
        "\n",
        "print(f\"Localized files selected from the `{vcf_column}` column in the Terra data table:\")\n",
        "# print(json.dumps(samples, indent=2))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "BSyfzlZogQ63",
        "outputId": "97de878c-5fad-48da-f1cb-6f16058790db"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Localized files selected from the `full_anno_pass_vcf` column in the Terra data table:\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# TEST\n",
        "if run_test:\n",
        "  _test_sample_id = \"SSC04202\" #\"SSC02602\" #\"SSC02806\"#\"SSC06504\"\n",
        "  _test_vcf_filename, _test_unfiltered_vcf_filename = localize_obj_if_needed(_test_sample_id)\n",
        "  print(_test_vcf_filename)"
      ],
      "metadata": {
        "id": "t49s6TZJSBXq",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "4562e736-45fb-4ca7-b130-77a0f9d3ecc2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "./pzm/SSC04202.vcf.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Explore molecularly validated variants"
      ],
      "metadata": {
        "id": "YgCo4BD3xd_M"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "## Localize data from a GS bucket\n",
        "validation_results_filename = os.path.join(working_dir, \"validation_results.txt\")\n",
        "\n",
        "if not isfile(validation_results_filename):\n",
        "  copy_object(\"gs://broad-dsde-methods-vj/mutec2-elise/Mutect_validation_results_20230227.txt\", validation_results_filename)\n",
        "\n",
        "molecularly_val_df = pd.read_csv(validation_results_filename, sep=\"\\t\")\n",
        "print(molecularly_val_df.shape)\n",
        "molecularly_val_df.head()"
      ],
      "metadata": {
        "id": "YMNidDxDLKOb",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 317
        },
        "outputId": "4debf956-751c-4b4b-a334-613eeda16280"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "(1635, 38)\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "   family_id participant_id chr   hg38_pos ref alt   ANY_PUB DOU_RESOLUTION  \\\n",
              "0      11000       11000.p1  13  101237830   G   A  germline       germline   \n",
              "1      11003       11003.p1   7   77942799   G   A    mosaic         mosaic   \n",
              "2      11003       11003.s1  12   56254338   C   A    mosaic         mosaic   \n",
              "3      11008       11008.p1  14   65629547   C   T    mosaic         mosaic   \n",
              "4      11008       11008.s1  16   16136537   G   A    mosaic         mosaic   \n",
              "\n",
              "  LIM_RESOLUTION FREED_RESOLUTION  ... lim_blood_tpcr_vaf freed_pyro_vaf  \\\n",
              "0       germline              NaN  ...                NaN            NaN   \n",
              "1            NaN              NaN  ...                NaN            NaN   \n",
              "2            NaN              NaN  ...                NaN            NaN   \n",
              "3            NaN              NaN  ...                NaN            NaN   \n",
              "4            NaN              NaN  ...                NaN            NaN   \n",
              "\n",
              "  krupp_exome_vaf krupp_wb_vaf krupp_lcl_vaf dou_exome_vaf dou_pasm_maf  \\\n",
              "0             NaN          NaN           NaN      0.324324        0.392   \n",
              "1             NaN          NaN           NaN      0.076087        0.079   \n",
              "2             NaN          NaN           NaN      0.327869        0.257   \n",
              "3             NaN          NaN           NaN      0.055556        0.071   \n",
              "4             NaN          NaN           NaN      0.193548        0.228   \n",
              "\n",
              "  WES_vaf_count   WES_vaf WES_vaf_source  \n",
              "0             2  0.306667  lim_exome_vaf  \n",
              "1             1  0.076087  dou_exome_vaf  \n",
              "2             1  0.327869  dou_exome_vaf  \n",
              "3             1  0.055556  dou_exome_vaf  \n",
              "4             1  0.193548  dou_exome_vaf  \n",
              "\n",
              "[5 rows x 38 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-937734d1-6e3f-48cf-92b3-ae970ea319ef\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>family_id</th>\n",
              "      <th>participant_id</th>\n",
              "      <th>chr</th>\n",
              "      <th>hg38_pos</th>\n",
              "      <th>ref</th>\n",
              "      <th>alt</th>\n",
              "      <th>ANY_PUB</th>\n",
              "      <th>DOU_RESOLUTION</th>\n",
              "      <th>LIM_RESOLUTION</th>\n",
              "      <th>FREED_RESOLUTION</th>\n",
              "      <th>...</th>\n",
              "      <th>lim_blood_tpcr_vaf</th>\n",
              "      <th>freed_pyro_vaf</th>\n",
              "      <th>krupp_exome_vaf</th>\n",
              "      <th>krupp_wb_vaf</th>\n",
              "      <th>krupp_lcl_vaf</th>\n",
              "      <th>dou_exome_vaf</th>\n",
              "      <th>dou_pasm_maf</th>\n",
              "      <th>WES_vaf_count</th>\n",
              "      <th>WES_vaf</th>\n",
              "      <th>WES_vaf_source</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>11000</td>\n",
              "      <td>11000.p1</td>\n",
              "      <td>13</td>\n",
              "      <td>101237830</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>germline</td>\n",
              "      <td>germline</td>\n",
              "      <td>germline</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.324324</td>\n",
              "      <td>0.392</td>\n",
              "      <td>2</td>\n",
              "      <td>0.306667</td>\n",
              "      <td>lim_exome_vaf</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>11003</td>\n",
              "      <td>11003.p1</td>\n",
              "      <td>7</td>\n",
              "      <td>77942799</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.076087</td>\n",
              "      <td>0.079</td>\n",
              "      <td>1</td>\n",
              "      <td>0.076087</td>\n",
              "      <td>dou_exome_vaf</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>11003</td>\n",
              "      <td>11003.s1</td>\n",
              "      <td>12</td>\n",
              "      <td>56254338</td>\n",
              "      <td>C</td>\n",
              "      <td>A</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.327869</td>\n",
              "      <td>0.257</td>\n",
              "      <td>1</td>\n",
              "      <td>0.327869</td>\n",
              "      <td>dou_exome_vaf</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>11008</td>\n",
              "      <td>11008.p1</td>\n",
              "      <td>14</td>\n",
              "      <td>65629547</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.055556</td>\n",
              "      <td>0.071</td>\n",
              "      <td>1</td>\n",
              "      <td>0.055556</td>\n",
              "      <td>dou_exome_vaf</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>11008</td>\n",
              "      <td>11008.s1</td>\n",
              "      <td>16</td>\n",
              "      <td>16136537</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.193548</td>\n",
              "      <td>0.228</td>\n",
              "      <td>1</td>\n",
              "      <td>0.193548</td>\n",
              "      <td>dou_exome_vaf</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 38 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-937734d1-6e3f-48cf-92b3-ae970ea319ef')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  \n",
              "    <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-937734d1-6e3f-48cf-92b3-ae970ea319ef')\"\n",
              "            title=\"Generate charts.\"\n",
              "            style=\"display:none;\">\n",
              "      \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "      <g>\n",
              "          <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "      </g>\n",
              "  </svg>\n",
              "    </button>\n",
              "   \n",
              "  <style>\n",
              "    .colab-df-quickchart {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-quickchart:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-quickchart {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-quickchart:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#df-937734d1-6e3f-48cf-92b3-ae970ea319ef button.colab-df-quickchart');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function quickchart(key) {\n",
              "        const containerElement = document.querySelector('#df-937734d1-6e3f-48cf-92b3-ae970ea319ef');\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'generateCharts', [key], {})\n",
              "      }\n",
              "    </script>    \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-937734d1-6e3f-48cf-92b3-ae970ea319ef button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-937734d1-6e3f-48cf-92b3-ae970ea319ef');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 16
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "molecularly_val_df.columns"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nL7PHMdAFPsR",
        "outputId": "724f7bb5-c026-4a4d-9121-19a279d66337"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Index(['family_id', 'participant_id', 'chr', 'hg38_pos', 'ref', 'alt',\n",
              "       'ANY_PUB', 'DOU_RESOLUTION', 'LIM_RESOLUTION', 'FREED_RESOLUTION',\n",
              "       'OROAK_RESOLUTION', 'relation', 'WES_ID', 'confirmed_wes_wgs_rel',\n",
              "       'WGS_id', 'filtered_zip_vcf', 'funcotate_vcf', 'cram', 'cram_index',\n",
              "       'WGS_IGV_cram', 'pos', 'Mutect_file', 'Mutect_AD', 'Mutect_VAF',\n",
              "       'lim_exome_vaf', 'lim_topo_vaf', 'lim_pyro_vaf', 'lim_cell_tpcr_vaf',\n",
              "       'lim_blood_tpcr_vaf', 'freed_pyro_vaf', 'krupp_exome_vaf',\n",
              "       'krupp_wb_vaf', 'krupp_lcl_vaf', 'dou_exome_vaf', 'dou_pasm_maf',\n",
              "       'WES_vaf_count', 'WES_vaf', 'WES_vaf_source'],\n",
              "      dtype='object')"
            ]
          },
          "metadata": {},
          "execution_count": 17
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Count per-sample loci and sort\n",
        "count_mol_val_df = molecularly_val_df.groupby(\"WGS_id\")[\"WGS_id\"].count().reset_index(name=\"count\").sort_values([\"count\"], ascending=False)\n",
        "print(count_mol_val_df)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IrIdBtEuxm-8",
        "outputId": "910d0ee1-6889-4117-9b07-7dd4aa6a115f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "       WGS_id  count\n",
            "427  SSC09045     39\n",
            "261  SSC05448     33\n",
            "99   SSC02667     26\n",
            "428  SSC09053     25\n",
            "201  SSC04423     24\n",
            "..        ...    ...\n",
            "171  SSC03905      1\n",
            "366  SSC07615      1\n",
            "365  SSC07605      1\n",
            "364  SSC07604      1\n",
            "312  SSC06504      1\n",
            "\n",
            "[625 rows x 2 columns]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Plot the counts.\n",
        "sns.set()\n",
        "fig, ax = plt.subplots(figsize=(6, 4))\n",
        "sns.histplot(data=count_mol_val_df[\"count\"], ax=ax, kde=False, bins=16)\n",
        "\n",
        "plt.xlabel(\"\\nMolecularly validated variants per sample\")\n",
        "plt.ylabel(\"Count\")\n",
        "plt.yscale(\"log\")\n",
        "plt.show()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 417
        },
        "id": "SRftCEv3x0f3",
        "outputId": "7584c370-d802-4b38-f286-34f058ac82b7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 600x400 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiUAAAGQCAYAAABxghCwAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA7mklEQVR4nO3deXhTZf7+8TtJW6CFFCqbwypoK1hKQaECWhArUBZRQEGGZQBZnAoO/uarMI4oI6MM48biLirLCLiACkIRXIqCAs4ojBsiFYQyFKTQnZYm5/cHkwyhW9qm5KR9v67LC3Py5Dyf5zwn6Z1zThKLYRiGAAAA/Mzq7wIAAAAkQgkAADAJQgkAADAFQgkAADAFQgkAADAFQgkAADAFQgkAADAFQgkAADCFIH8XECgMw5DTWf73zFmtFq/aBSrGF7hq8tgkxhfIavLYpJo9Pm/GZrVaZLFYvFofocRLTqehjIzcMtsEBVnVqFGYsrLyVFTkvEiVXTyML3DV5LFJjC+Q1eSxSTV7fN6OLSIiTDabd6GE0zcAAMAUCCUAAMAUCCUAAMAUCCUAAMAUCCUAAMAUCCUAAMAUCCUAAMAUCCUAAMAUCCUAAMAUCCUAAMAUCCUAAMAU+O0bP7NaLbJavftNgKpwOr37QUEAAPyFUOJHVqtFDRuFymat/gNWDqdTp0/lEUwAAKZFKPEjq9Uim9Wq15O/1/GMvGrrp2lEqEYP6FCjfz4bABD4CCUmcDwjT2kncvxdBgAAfsWFrgAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBQIJQAAwBSC/F1AdTl16pT+7//+T4cPH1ZISIiio6M1d+5chYSE+Ls0AABQghp7pMRisWjq1KnavHmz3n33XRUUFGjlypX+LgsAAJTCdKHk0KFDmjNnjoYOHaqOHTtq8ODBJbY7cOCAJkyYoNjYWPXq1UsLFixQYWGh+/6GDRuqW7dukiSr1aro6GgdPXr0oowBAABUnOlCyf79+5WSkqI2bdqoffv2JbbJzMzU+PHjdfbsWS1evFgzZ87UG2+8ofnz55fYvqCgQOvWrVN8fHx1lg4AAKrAdNeU9O3bVwkJCZKkWbNm6ZtvvinWZvXq1crNzdWSJUvUsGFDSZLD4dDcuXM1depUNWvWzN3W6XTq/vvvV1xcHKEEAAATM92REqu1/JK2bdumHj16uAOJJCUmJsrpdGr79u0ebefOnSur1ao//elPvi4VAAD4kOmOlHgjNTVVw4cP91hmt9vVpEkTpaamupctWLBAx44d05IlS7wKO+UJCip7HTab1ePf8rjaWSwWWSyWqhVXBte6va2rNBUdX6CpyeOryWOTGF8gq8ljk2r2+KpjbAEZSrKysmS324stDw8PV2ZmpqRz16YsXbpU7dq104gRIyRJPXv21P3331+pPq1Wixo1CvOqrd1er0LrttmsCgqyVaYsr9cvVbyu0vhqPWZVk8dXk8cmMb5AVpPHJtXs8flybAEZSrxxxRVXaN++fT5bn9NpKCsrr8w2NptVdns9ZWXly+FwlrtOV3uHw6miIoevSi3GVYu3dZWmouMLNDV5fDV5bBLjC2Q1eWxSzR6ft2Oz2+t5fTQlIEOJ3W5XdnZ2seWZmZkKDw+vtn6Lirzboc6FDO93PsMwZBhGZcvyav1Sxesqja/WY1Y1eXw1eWwS4wtkNXlsUs0eny/HFpAnudq1a+dx7YgkZWdn68SJE2rXrp2fqgIAAFURkKEkPj5eO3bsUFZWlntZcnKyrFarevXq5cfKAABAZZnu9E1+fr5SUlIkSWlpacrJyVFycrIkqXv37oqIiNCoUaO0YsUKJSUlaerUqUpPT9eCBQs0atQoj+8oAQAAgcN0oeTkyZO65557PJa5bi9fvlxxcXEKDw/XsmXL9MgjjygpKUlhYWEaMWKEZs6c6Y+SAQCAD5gulLRs2dKrT820b99er732WvUXBAAALoqAvKYEAADUPIQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCoQSAABgCkH+LgAXj81WtQzqenxZ63E6DTmdRpX6AQDUToSSWqBBaLCcTkN2ez2frK+s9TicTp0+lUcwAQBUGKGkFqhbJ0hWq0WrNv+g9JO5lV6PxWKRzWaVw+GUYRQPHU0jQjV6QAdZrRZCCQCgwggltcjxjDylncip9OMtFouCgmwqKnKUGEoAAKgKLnQFAACmQCgBAACmQCgBAACmQCgBAACmQCgBAACmQCgBAACmUOlQMm7cOH3++eel3v/FF19o3LhxlV09AACoZSodSnbt2qVff/211PszMjK0e/fuyq4eAADUMlU6fWOxWEq979ChQwoLC6vK6gEAQC1SoW90XbdundatW+e+/dxzz+mNN94o1i47O1v79u1TfHx81SsEAAC1QoVCSX5+vk6dOuW+nZubK6u1+MGW0NBQjRo1SklJSVWvEAAA1AoVCiWjR4/W6NGjJUl9+/bVAw88oBtvvLFaCgMAALVLpX+Q76OPPvJlHQAAoJar8q8E5+Tk6OjRo8rKyirxl2O7detW1S4AAEAtUOlQkpGRoXnz5umDDz6Qw+Eodr9hGLJYLPr++++rVCAAAKgdKh1K5syZo48//lhjx47VNddcI7vd7su6AABALVPpULJ9+3aNHz9e9913ny/rAQAAtVSlvzytbt26atGihS9rAQAAtVilQ8nNN9+srVu3+rIWAABQi1X69E3//v21e/duTZo0SSNHjlTz5s1ls9mKtbvqqquqVCAAAKgdKh1KXF+iJkk7duwodj+fvgEAABVR6VDy2GOP+bIOAABQy1U6lNx6662+rAMAANRylb7QFQAAwJcqfaRk9uzZ5baxWCx69NFHK9sFApTNdnGyrtNpyOks/tMGAIDAVOlQsnPnzmLLnE6nTpw4IYfDoYiICNWrV69KxSGwNAgNltNpyG6/OPPucDp1+lQewQQAagif/0rw2bNntWbNGi1btkyvvPJKpQtD4KlbJ0hWq0WrNv+g9JO51dpX04hQjR7QQVarhVACADVElX8l+ELBwcEaM2aMfvrpJz3yyCN68cUXfd0FTO54Rp7STuT4uwwAQICptpP/V155pXbv3l1dqwcAADVMtYWSHTt2cE0JAADwWqVP3yxZsqTE5dnZ2dq9e7e+++47TZkypdKFAQCA2sXnoSQ8PFytWrXS3Llzdfvtt1e6MAAAULtUOpT88MMPvqwDAADUcnyjKwAAMIUqfyR4165d+uSTT3T06FFJ0m9+8xv16dNH3bt3r3JxAACg9qh0KCksLNT/+3//T1u3bpVhGLLb7ZKkrKwsvfrqq7rpppv0xBNPKDg42GfFAgCAmqvSp2+eeeYZbdmyRRMmTNBnn32mXbt2adeuXdq+fbsmTpyoDz74QM8884wvawUAADVYpUPJ+vXrdeutt+q+++5T48aN3csvueQS/d///Z9uueUWvffeez4pEgAA1HyVDiUnTpxQTExMqffHxMToxIkTlV09AACoZSodSpo3b65du3aVev/u3bvVvHnzyq4eAADUMpUOJbfccos2bdqkOXPmKDU1VQ6HQ06nU6mpqXrooYeUnJysW2+91Ze1AgCAGqzSn76ZNm2aDh8+rDfeeENvvvmmrNZz+cbpdMowDN16662aNm2azwoFAAA1W6VDic1m0/z58/W73/1O27ZtU1pamiSpRYsWio+P15VXXumzIgEAQM1XoVBSUFCgv/71r7riiis0duxYSdKVV15ZLIAsX75cq1ev1gMPPMD3lAAAAK9U6JqSNWvWaN26derTp0+Z7fr06aO3335bb775ZlVqAwAAtUiFQsmmTZvUr18/tWrVqsx2rVu31oABA/T+++9XqTgAAFB7VCiU/Pjjj7r66qu9atulSxft27evUkUBAIDap0Kh5OzZs15fIxIcHKzCwsJKFQUAAGqfCoWSpk2bav/+/V613b9/v5o2bVqpogAAQO1ToVDSs2dPvfvuuzp58mSZ7U6ePKl3331XPXv2rFJxAACg9qhQKJk8ebIKCgo0fvx47dmzp8Q2e/bs0e9+9zsVFBTozjvv9EmRAACg5qvQ95S0atVKTz/9tO69916NGjVKrVq1UmRkpMLCwpSbm6v9+/frl19+Ud26dfXkk0+qdevW1VU3AACoYSr8ja59+vTRe++9p5deekmffPKJtm7d6r6vadOmuu222zR58uRyPzYMAABwvkp9zXzLli01d+5cSVJOTo5yc3MVFham+vXr+7Q4AABQe1T6t29c6tevTxgBAABVVqELXQEAAKoLoQQAAJgCoQQAAJhCjQ4lc+bM0fXXX6+oqCh/lwIAAMpRo0PJkCFDtG7dOn+XAQAAvGC6UHLo0CHNmTNHQ4cOVceOHTV48OAS2x04cEATJkxQbGysevXqpQULFhT7AcBu3bqpcePGF6NsAABQRVX+SLCv7d+/XykpKercubOcTqcMwyjWJjMzU+PHj1fbtm21ePFipaena/78+Tpz5ozmzJnjh6oBAEBVmS6U9O3bVwkJCZKkWbNm6ZtvvinWZvXq1crNzdWSJUvUsGFDSZLD4dDcuXM1depUNWvW7GKWDAAAfMB0ocRqLf+M0rZt29SjRw93IJGkxMREPfTQQ9q+fbuGDRtWLbUFBZVdm81m9fi3PK52FotFFoulasWVwb1ui6rWj+V//1pUfD0+68ebUv67/uBgm9fbuzxWa8nrdDqNEo/YBZKK7puBhvEFrpo8Nqlmj686xma6UOKN1NRUDR8+3GOZ3W5XkyZNlJqaWi19Wq0WNWoU5lVbu71ehdZts1kVFGSrTFnerf+/Qc9m9U0/QbaS1+HrfsoS3qCOnE5D9evX9fm6L1yn02m4A0ugq+i+GWgYX+CqyWOTavb4fDm2gAwlWVlZstvtxZaHh4crMzPTfXvWrFnasWOHJCk+Pl5xcXH6+9//Xqk+nU5DWVl5Zbax2ayy2+spKytfDoez3HW62jscThUVOSpVlzccTqf73yr1YzkXSIocDqmEAwc+68cLIUFWWa0Wrdr8g45nlD0vXrOcC1QOp9M9vqYRobqj/5Vez6lZVXTfDDSML3DV5LFJNXt83o7Nbq/n9dGUgAwl3po/f75P11dU5N0OdS5keL/zGUb1nh5wr9tQlfpxn7IpZT2+6scbrvUfz8jTkePZPlmnxWJRUJBNRUUO9/pd/1Z0Ts2qpoyjNIwvcNXksUk1e3y+HFtAnuSy2+3Kzi7+hygzM1Ph4eF+qAgAAFRVQIaSdu3aFbt2JDs7WydOnFC7du38VBUAAKiKgAwl8fHx2rFjh7KystzLkpOTZbVa1atXLz9WBgAAKst015Tk5+crJSVFkpSWlqacnBwlJydLkrp3766IiAiNGjVKK1asUFJSkqZOnar09HQtWLBAo0aN4jtKAAAIUKYLJSdPntQ999zjscx1e/ny5YqLi1N4eLiWLVumRx55RElJSQoLC9OIESM0c+ZMf5QMAAB8wHShpGXLltq3b1+57dq3b6/XXnut+gsCAAAXRUBeUwIAAGoeQgkAADAFQgkAADAFQgkAADAFQgkAADAFQgkAADAF030kGKjNrFaLrFaLz9fr+oVO179OpyGns3p/NBEAKopQApiE1WpRw0ahslmr7wCm3V5PkuRwOnX6VB7BBICpEEoAk7BaLbJZrXo9+Xsdz8jz6botFotsNqscDqeaNKqn0QM6yGq1EEoAmAqhBDCZ4xl5SjuR49N1WiwWBQXZVFTkkGEQRACYExe6AgAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUyCUAAAAUwjydwFAILDZqj+/X4w+AMDMCCVAGRqEBsvpNGS31/N3KQBQ4xFKgDLUrRMkq9WiVZt/UPrJ3GrtK6pthBJ7XiaLxVKt/QCAWRFKAC8cz8hT2omcau2jSSOOxgCo3TiJDQAATIFQAgAATIFQAgAATIFQAgAATIFQAgAATIFQAgAATIFQAgAATIFQAgAATIFQAgAATIFvdAWACrBaLbJaL85PATidhpxO46L0BZgBoQQAvGS1WtSwUahs1otzkNnhdOr0qbyL0hdgBoQSAPCS1WqRzWrV68nf63hG9YaFphGhGj2gw0U7KgOYAaEEACroYvxAI1AbcaErAAAwBUIJAAAwBUIJAAAwBUIJAAAwBUIJAAAwBUIJAAAwBUIJAAAwBUIJAAAwBUIJAAAwBUIJAAAwBUIJAAAwBUIJAAAwBUIJAAAwBUIJAAAwhSB/FwAAvmKzVe/7rOpef3l9Vlf/Tqchp9OolnUDFUEoARDwLBaLnE5Ddns9f5fiMw1Cg4uNqbrG53A6dfpUHsEEfkcoARDwrFaLrFaLVm3+Qeknc6utn6i2EUrseZksFku19eFSt06Qe0zHM/Jks1nlcDhlGL4NDk0jQjV6QAdZrRZCCfyOUAKgxjiekae0EznVtv4mjS7+kRjXmIKCbCoqcvg8lABmwoWuAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFAglAADAFGp0KDlw4ICGDx+u/v37a9y4cTp+/Li/SwIAAKWo0aHkoYce0l133aXNmzfrxhtv1BNPPOHvkgAAQClMF0oOHTqkOXPmaOjQoerYsaMGDx5cYrsDBw5owoQJio2NVa9evbRgwQIVFha67//111918OBBJSQkSJJGjBihLVu2XJQxAACAigvydwEX2r9/v1JSUtS5c2c5nU4ZhlGsTWZmpsaPH6+2bdtq8eLFSk9P1/z583XmzBnNmTNHknTs2DFdeuml7seEhYWpTp06OnXqlBo1anTRxgMAALxjulDSt29f99GNWbNm6ZtvvinWZvXq1crNzdWSJUvUsGFDSZLD4dDcuXM1depUNWvW7GKWDAAAfMB0ocRqLf+M0rZt29SjRw93IJGkxMREPfTQQ9q+fbuGDRum5s2b6z//+Y/7/tzcXBUUFFTpKElQUNm12WxWj3/L42pnsVhksVgqXVd53Ou2qGr9WP73r0XF1+OzfrwppTr6KmF8AT8m98r/969r3cHBNq/31aowDKmaN93/npvVPE/+2h/Ke+75op+LsS+UpKKvm96yWCyyWqt5x5PkdBolHtF38eX4zDIml+qYO9OFEm+kpqZq+PDhHsvsdruaNGmi1NRUSVLjxo3Vpk0bbd26VQkJCXrrrbfcR2Aqw2q1qFGjMK/a2u31KrRum82qoCBbZcrybv3/DXo2q2/6CbKVvA5f91OW6uzr/PHVlDG5BNlsCm9QR06nofr161ZLHxdyOo2L8kIqVf88+W1/+O8+Wdpzr0r9/PcPSkVft3zN1/1frP3O2358MT6zjcnFl3MXkKEkKytLdru92PLw8HBlZma6bz/88MO6//77tWDBAjVr1kyPP/54pft0Og1lZeWV2cZms8pur6esrHw5HM5y1+lq73A4VVTkqHRt5XE4ne5/q9SP5dyLYpHDIZUQon3Wjxeqpa8SxhfwY3I5b2whQVZZrRat2vyDjmeUvU9XVVSbRhrQ87Jq78vVT3XPk9/2B4ejzOdelfr572uVt69bvlbR182KrLO697umEaG6o/+VZdbuq/GZaUwX1lReW7u9ntdHUwIylHjriiuu0Nq1a322vqIi73aocyHD+53PMLw7VFZZ7nUbqlI/7sPGpazHV/14ozr6Kml8gT4ml5LGdjwjT0eOZ/u0nws1blj3ovTVpNH/3qkFwnOpon25g0g19OtaX0Vft3ytOvpPP5mrtBM5Pl3n+Sqy7Xw1PjONycWXc2e6jwR7w263Kzu7+AtcZmamwsPD/VARAACoqoAMJe3atXNfO+KSnZ2tEydOqF27dn6qCgAAVEVAhpL4+Hjt2LFDWVlZ7mXJycmyWq3q1auXHysDAACVZbprSvLz85WSkiJJSktLU05OjpKTkyVJ3bt3V0REhEaNGqUVK1YoKSlJU6dOVXp6uhYsWKBRo0bxHSUAAAQo04WSkydP6p577vFY5rq9fPlyxcXFKTw8XMuWLdMjjzyipKQkhYWFacSIEZo5c6Y/SgYAAD5gulDSsmVL7du3r9x27du312uvvVb9BQEAgIsiIK8pAQAANQ+hBAAAmAKhBAAAmAKhBAAAmILFqO7vSq4hDMOQ0+ndryZW5PcNbDarMnMKVFSNvzkREmxTg9AQn/RjkUVGKT++4ct+ylNdfV04vpowJhfX2GrSmFzqBNtU/yL0489tV9ZzryqCbFaF16/jl9+9cano66a366zuefJ22/lqfGYa0/k1ldfWarV4/avahBIAAGAKnL4BAACmQCgBAACmQCgBAACmQCgBAACmQCgBAACmQCgBAACmQCgBAACmQCgBAACmQCgBAACmQCgBAACmQCgBAACmQCgBAACmQCgBAACmQCjxkQMHDmjChAmKjY1Vr169tGDBAhUWFvq7rCpbu3atoqKiiv33+OOP+7u0Sjl06JDmzJmjoUOHqmPHjho8eHCJ7d588031799fnTp10s0336yPP/74Ildacd6MbezYsSXO54EDB/xQsfc2bdqku+66S/Hx8YqNjdXQoUP11ltv6cIfOQ/EeZO8G1+gzp0kpaSkaMyYMbr22msVHR2tG2+8UY899piys7M92n300Ue6+eab1alTJ/Xv319vv/22nyr2njdjmzVrVolzt23bNj9WXjm5ubmKj49XVFSU/v3vf3vc54vnX5CvCq3NMjMzNX78eLVt21aLFy9Wenq65s+frzNnzmjOnDn+Ls8nXn75ZTVo0MB9u1mzZn6spvL279+vlJQUde7cWU6ns9gfNUl6//339eCDD2ratGm69tprtXHjRt199936xz/+odjY2ItftJe8GZskde3aVffff7/HspYtW16MEivttddeU4sWLTRr1iw1atRIO3bs0IMPPqhjx47p7rvvlhS48yZ5Nz4pMOdOkk6fPq2YmBiNHTtWDRs21P79+7V48WLt379fr7zyiiTpyy+/1N13360RI0boT3/6k7744gs98MADCgsL04ABA/w8gtJ5MzZJatWqVbE3c+3bt7/Y5VbZs88+K4fDUWy5z55/Bqrs+eefN2JjY41Tp065l61evdro0KGDcezYMf8V5gNvv/22ERkZaZw8edLfpfiEw+Fw///9999vDBo0qFibfv36Gffee6/HspEjRxp33nlntddXFd6MbcyYMcaUKVMuZlk+UdL+9+c//9no2rWre9yBOm+G4d34AnXuSrNmzRojMjLS/Ro5ceJEY+TIkR5t7r33XiMxMdEf5VXJhWMr7fkYaH766ScjNjbWWLVqlREZGWns3bvXfZ+vnn+cvvGBbdu2qUePHmrYsKF7WWJiopxOp7Zv3+6/wlCM1Vr2Ln/48GEdPHhQiYmJHssHDhyozz//3NSn5MobWyCLiIgotqxDhw7KyclRXl5eQM+bVP74aiLX6+XZs2dVWFionTt3FjsiMnDgQB04cEBHjhzxQ4WVd/7YapJ58+Zp1KhRuuyyyzyW+/L5V3NfxS6i1NRUtWvXzmOZ3W5XkyZNlJqa6qeqfGvw4MHq0KGDbrzxRr3wwgslHr6rCVzzdeGTrn379jp79qwOHz7sj7J8ateuXYqNjVWnTp00ZswY7d69298lVco///lPNWvWTPXr16+R83b++FwCfe4cDocKCgr07bff6plnnlHfvn3VsmVL/fLLLzp79myx11HX6Y1AeB0tbWwuhw4d0tVXX63o6GgNGzZMW7du9WO1FZecnKwff/xRSUlJxe7z5fOPa0p8ICsrS3a7vdjy8PBwZWZm+qEi32nSpImmT5+uzp07y2Kx6KOPPtLTTz+t9PT0GnO9zPlc83XhfLpuB/p8duvWTUOHDlXbtm11/PhxLV26VBMmTNCKFSvUpUsXf5fntS+//FIbN250X19R0+btwvFJNWPubrjhBqWnp0uSrr/+ej3xxBOSasb8lTY26dxRr06dOunyyy9Xdna2Vq1apaSkJC1cuNDU18u45Ofna/78+Zo5c6ZHSHbx5fwRSlCm66+/Xtdff7379nXXXac6depo2bJlmjZtmpo2berH6lBRM2bM8Ljdp08fDR48WM8++6xeeuklP1VVMceOHdPMmTMVFxencePG+bscnyttfDVh7l588UXl5+frp59+0nPPPadp06bp1Vdf9XdZPlHa2Gw2m8aPH+/Rtm/fvho1apQWLVoUEKHkueee0yWXXKLhw4dXe1+cvvEBu91e7KNt0rl0GB4e7oeKqldiYqIcDoe+//57f5fic675unA+s7KyPO6vKUJDQ9W7d299++23/i7FK1lZWZo8ebIaNmyoxYsXu6+jqSnzVtr4ShJocydJV155pbp06aLbbrtNzz77rHbu3KktW7bUiPkrbWwlsVqt6tevnw4cOKAzZ85c5EorJi0tTa+88opmzJih7OxsZWVlua9zysvLU25urk/njyMlPtCuXbti5zyzs7N14sSJYudIYW6u+brwOqHU1FQFBwerVatW/iqt1jtz5oymTp2q7OxsrVmzxuMj6jVh3soaX00UFRWl4OBg/fLLL+rbt6+Cg4OVmprqcWTW9boaaK+j548t0B05ckRnz57VlClTit03btw4de7c2X2qyhfPP46U+EB8fLx27NjhToXSuYuCrFarevXq5cfKqsfGjRtls9nUsWNHf5fic61atVLbtm2VnJzssXzjxo3q0aOHQkJC/FRZ9cjLy9Mnn3yiTp06+buUMhUVFekPf/iDUlNT9fLLLxf7npxAn7fyxleSQJm70uzZs0dnz55Vy5YtFRISori4OG3evNmjzcaNG9W+ffuA+C6W850/tpI4nU4lJyfriiuuUN26dS9ydRXToUMHLV++3OO/2bNnS5Lmzp2rhx56yKfPP46U+MCoUaO0YsUKJSUlaerUqUpPT9eCBQs0atSogP2SMZdJkyYpLi5OUVFRkqQPP/xQb7zxhsaNG6cmTZr4ubqKy8/PV0pKiqRzhyVzcnLcT6Tu3bsrIiJC06dP1x//+Ee1bt1acXFx2rhxo/bu3auVK1f6s/RylTc21x+8m266SS1atNDx48f16quv6sSJE1q4cKE/Sy/X3Llz9fHHH2vWrFnKycnR119/7b6vY8eOCgkJCdh5k8of3969ewN27iTp7rvvVnR0tKKiolS3bl398MMPWrp0qaKiopSQkCBJuuuuuzRu3Dg9/PDDSkxM1M6dO7VhwwY99dRTfq6+bOWNLS0tTbNmzdKgQYPUpk0bZWZmatWqVfrmm2+0ePFif5dfLrvdrri4uBLvu+qqq3TVVVdJks+efxbDKOVrH1EhBw4c0COPPKKvvvpKYWFhGjp0qGbOnGn6d2jlmTdvnj799FMdO3ZMTqdTbdu21W233aaxY8fKYrH4u7wKO3LkiG688cYS71u+fLn7yffmm2/qpZde0tGjR3XZZZfp3nvv1Q033HAxS62w8sbWvHlz/eUvf9G+fft0+vRp1atXT126dNHdd9+tmJiYi1xtxfTt21dpaWkl3vfhhx+635EG4rxJ5Y/P4XAE7NxJ5y4C3bhxo3755RcZhqEWLVropptu0qRJkzw+zfHhhx/q6aef1s8//6zf/OY3mjJlikaMGOHHystX3thOnz6t2bNn67vvvtPJkycVHBys6OhoTZkyxeNUVSDZuXOnxo0bp7feesvjSJ0vnn+EEgAAYApcUwIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAIAAEyBUAK/6du3r2bNmuXvMrR27VpFRUXpyJEjPl3v2LFjNXbsWJ+u05eioqK0ePFi9+2KbAezzN35Fi9erKioKH+X4cHs+wB8y4z7YKAhlKBErj9QUVFR+vLLL4vdbxiGevfuraioKE2dOtUPFSIQpaena/Hixfr+++/9XUqNkZKS4hEugUBGKEGZ6tSpow0bNhRbvmvXLh07dkwhISF+qArVYejQodq7d69atGhRbX0cP35cS5YsqTWhZOnSpVq6dGm19pGSkqIlS5ZUax/AxUIoQZl69+6t5ORkFRUVeSzfsGGDrrrqKjVp0sRPlZlXfn6+v0uoFJvNpjp16shisfi7lIDn2gdCQkII7qXIy8vzdwkwIUIJyjRo0CCdPn1a27dvdy8rLCzU5s2bNWTIkBIfk5eXp/nz56t3796Kjo5W//79tXTpUhmGUW5/WVlZ+utf/+p+7E033aQXX3xRTqfTo53T6dSyZcs0ZMgQderUSddee60mTZqkf//735KkI0eOKCoqSmvXri3Wx4XXUpRk69atmjJliq677jpFR0crISFBzzzzjBwOh0e7sWPHavDgwfrmm2/029/+Vp07d9aTTz5ZbH25ubmKjY3VvHnzit137NgxdejQQS+88EKJtZw9e1bdu3fX7Nmzi92Xk5OjTp066W9/+5ukc3OzcOFCDRs2TFdffbViY2M1evRoffHFF2WOVyr5mhLDMPTss88qPj5enTt31tixY7V///5ijz19+rT+9re/aciQIerSpYu6du2qO++8Uz/88IO7zc6dOzVixAhJ0uzZs92nB8+foz179mjSpEm6+uqr1blzZ40ZM0b//Oc/i/X35Zdfavjw4erUqZMSEhK0evXqcscnSX/5y1/UpUuXEoPjvffeq169ernn2Bf7wIXXlHg7P679d+nSpVqzZo0SEhIUHR2t4cOHa+/eve52s2bN0j/+8Q9Jcm/P869peP/99zVs2DD3nAwZMkTLli0rcxud3/drr72mG264QTExMRozZox+/PHHYu0PHDigGTNmqHv37urUqZOGDRumDz/80KONa9/atWuXHn74YfXo0UO9e/cus44VK1Zo0KBB6ty5s7p166Zhw4Zp/fr17vvT0tL08MMPq3///oqJiVFcXJxmzJhR7JooV99ffvml5s2bp2uvvVbXXHON5syZo8LCQmVlZem+++5Tt27d1K1bNy1YsMDjtaqi26Mk7777roYNG6aYmBh1795dM2fO1H/+8x+vHlvbBPm7AJhbixYtFBsbq/fff9/9IrJt2zZlZ2dr4MCBWrFihUd7wzB01113uf8AdejQQZ9++qkWLFig9PR0/elPfyq1r/z8fI0ZM0bp6ekaNWqULr30Un311Vd68skndeLECT3wwAPutg888IDWrl2r+Ph4jRgxQg6HQ19++aX27NmjTp06VXnc69atU2hoqCZMmKDQ0FB98cUXWrRokXJycnT//fd7tD19+rQmT56sQYMG6eabb9Yll1xSbH1hYWFKSEjQpk2bNHv2bNlsNvd9GzZskGEYpYa84OBgJSQkaMuWLZo7d67HO++tW7eqsLBQAwcOlHQupLz55psaPHiwbrvtNuXm5uqtt97SnXfeqTfffFMdOnSo0HZYuHChnnvuOfXu3Vu9e/fWt99+q4kTJ+rs2bMe7Q4fPqytW7dqwIABatmypX799VetWbNGY8aM0fvvv69mzZqpffv2mjFjhhYtWqSRI0fq6quvliR17dpVkvT5559r8uTJio6O1t133y2LxaK1a9dq/Pjxev311xUTEyNJ2rdvnyZNmqSIiAhNnz5dRUVFWrx4cYnb/UIDBw7UP/7xD33yySdKTEx0L8/Pz9fHH3+sW2+91T03vt4HpIrPz4YNG5Sbm6uRI0fKYrHo5Zdf1vTp07V161YFBwdr5MiROn78uLZv364FCxZ4PHb79u2699571aNHD/3xj3+UJKWmpupf//qXxo8fX+62euedd5Sbm6vRo0eroKBAK1as0Pjx47V+/Xo1btxYkrR//37dcccdatasmSZPnqzQ0FBt2rRJSUlJWrx4sW666SaPdc6dO1cRERFKSkoq80jJG2+8oXnz5ql///4aN26cCgoKtG/fPu3Zs8f9PPn3v/+tr776SoMGDVLz5s2VlpamVatWady4cXr//fdVr149j3XOmzdPjRs31vTp07Vnzx6tWbNGDRo00FdffaVLL71UM2fO1LZt27R06VJFRkbqlltuqfD2KMlzzz2nhQsXKjExUSNGjFBGRoZWrlyp3/72t3rnnXdkt9vLnYtaxQBK8PbbbxuRkZHG3r17jZUrVxpdunQx8vPzDcMwjBkzZhhjx441DMMwbrjhBmPKlCnux23ZssWIjIw0nn32WY/1TZ8+3YiKijIOHTrkXnbDDTcY999/v/v2M888Y8TGxho///yzx2Mff/xxo0OHDsbRo0cNwzCMzz//3IiMjDQeeeSRYnU7nU7DMAzj8OHDRmRkpPH2228XaxMZGWksWrSo2FgPHz7sXuYa6/kefPBBo3PnzkZBQYF72ZgxY4zIyEhj1apVxdqPGTPGGDNmjPv2p59+akRGRhopKSke7YYMGeLRriSux3700UceyydPnmzceOON7ttFRUUe9RmGYWRmZho9e/Y0Zs+e7bG8vO1w8uRJ46qrrjKmTJni3q6GYRhPPvmkERkZ6TF3BQUFhsPh8Fj/4cOHjejoaGPJkiXuZXv37i1xXpxOp9GvXz9j4sSJHn3l5+cbffv2NSZMmOBe9vvf/97o1KmTkZaW5l72008/GR06dDAiIyMv3HTF+rn++uuN6dOneyzfuHGjERkZaezevduj7wtVdR/wdn5c+2/37t2N06dPu5dv3bq12H4wd+7cEsc9b948o2vXrkZRUVGJ26I0rr5jYmKMY8eOuZfv2bPHiIyMNB599FH3svHjxxuDBw/2GJPT6TRGjhxp9OvXz73MtW/dcccdXtVz1113GYMGDSqzTUnz89VXXxmRkZHGunXrivV94b41cuRIIyoqypgzZ457WVFRkREfH+8xZxXZHosWLfKYiyNHjhgdOnQwnnvuOY869+3bZ3Ts2LHYchgGp29QrsTERBUUFOjjjz9WTk6OPvnkk1Lf1W/btk02m63YxyAnTpwowzC0bdu2UvtJTk7W1VdfLbvdroyMDPd/PXv2lMPh0O7duyVJH3zwgSwWi+6+++5i6/DV9RB169Z1/39OTo4yMjJ0zTXXKD8/X6mpqR5tQ0JCNGzYsHLX2bNnTzVt2tTjEPSPP/6offv26eabby7zsddee60aNWqkjRs3updlZmZqx44d7qMk0rnrQlxHUpxOp06fPq2ioiJFR0fru+++K7fG8+3YsUNnz57VmDFjPLZrSe+yQ0JCZLWeezlxOBw6deqUQkNDddlll3nV7/fff6+DBw9qyJAhOnXqlHvu8/Ly1KNHD+3evVtOp1MOh0OfffaZEhIS9Jvf/Mb9+Pbt2+u6664rtx+LxaIBAwYoJSVFubm57uWbNm1Ss2bN3EdvpOrZByo6PwMHDlR4eLj79jXXXCPp3JGp8tjtduXn53uceq2IhIQENWvWzH07JiZGnTt3VkpKiqRzR4e++OILJSYmurdPRkaGTp06peuuu04HDx5Uenq6xzpvv/12j6OEZdV+7Ngxj1NVFzp/fs6ePatTp06pdevWstvtJW7LESNGeOzHMTExMgzDfUpROjc/0dHRJW7f8rZHSbZs2SKn06nExESP17TGjRurTZs22rlzZ+kboZbi9A3KFRERoR49emjDhg06c+aMHA6H+vfvX2LbtLQ0NW3aVPXr1/dY3r59e/f9pTl06JD27dunHj16lHh/RkaGJOmXX35R06ZN1bBhw0qMxjv79+/X008/rS+++EI5OTke92VnZ3vcbtasmVcXM1qtVg0ZMkSrVq1Sfn6+6tWrp/Xr16tOnToaMGBAmY8NCgpSv379tGHDBhUWFiokJEQffPCBzp496xFKpHOnHV555RX9/PPPHqdZWrZsWW6N5zt69KgkqW3bth7LIyIiPP5QSuf+wC5fvlyvv/66jhw54nHdhTfzdPDgQUkqdlrkfNnZ2SosLNSZM2fUpk2bYvdfdtllZf6BcBk4cKCWLVumjz76SEOGDFFubq5SUlLcp0hcqmMfkCo2P5deeqnHbdd2z8rKKref0aNHa9OmTZo8ebKaNWumXr16KTExUfHx8V7VWdI2btu2rTZt2iTp3PPQMAwtXLhQCxcuLHEdJ0+e9PhD7u0+OHnyZO3YsUO33Xab2rRpo169emnw4MEeofHMmTN64YUXtHbtWqWnp3tcB3Lh/EjyCLGS1KBBA0nFt3GDBg2UmZlZ7PHlbY+SHDx4UIZhqF+/fiXeHxTEn+ALsUXglcGDB+vBBx/Ur7/+qvj4+Go5D+p0OtWrVy/deeedJd5/4R/HspR2xOTCixRLkpWVpTFjxqh+/fqaMWOGWrdurTp16ujbb7/V448/Xuyi2/PfsZXnlltu0dKlS7V161YNHjxYGzZsUJ8+fdwvkGUZNGiQ1qxZo23btikhIUHJyclq166drrzySnebd999V7NmzVJCQoImTZqkSy65RDabTS+88IJX764r6/nnn9fChQs1fPhw3XPPPQoPD5fVatWjjz7q1QXOrjb33Xdfqde9hIaGqrCwsMq1xsbGqkWLFtq0aZOGDBmijz/+WGfOnPEId9W1D1R0fko7quDNNr3kkkv0zjvv6LPPPtO2bdu0bds2rV27Vrfccov7wuiqcG2DiRMn6vrrry+xTevWrT1u16lTx6t1t2/fXsnJyfrkk0/06aef6oMPPtDrr7+upKQkzZgxQ5L0yCOPuK85io2NVYMGDWSxWDRz5swSt4/rSJ63y33B6XTKYrHopZdeKnEuQ0NDq63vQEUogVduuukmPfTQQ/r666/11FNPldquRYsW+vzzz5WTk+NxtMR1uLus78Bo3bq18vLy1LNnzzJrad26tT777DOdPn261Hfhpb2jdL37L8uuXbt0+vRpLVmyRN26dXMv98U3vkZGRqpjx45av369mjdvrqNHj+rPf/6zV4/t1q2bmjRpoo0bN6pr16764osvNG3aNI82mzdvVqtWrbRkyRKPYLZo0aIK1+p6Z3nw4EG1atXKvTwjI6PYO8nNmzcrLi5Ojz76qMfyrKwsNWrUyH27tLDoWn/9+vXLnP+IiAjVrVtXhw4dKnbfzz//XM6I/icxMVHLly9XTk6ONm7c6L6g26W69gFfzo9LWacsQ0JC1LdvX/Xt21dOp1MPP/yw1qxZo9///vclvvM/X0nb+ODBg+7nsGvOgoODy33OVkZoaKgGDhyogQMHqrCwUNOnT9fzzz+vqVOnqk6dOtq8ebNuueUWj28WLigoKPEoiS+Utz1K0rp1axmGoZYtW+qyyy6rlrpqGq4pgVfCwsL08MMPa/r06erbt2+p7eLj4+VwONwfU3R57bXXZLFYyjx0nJiYqK+++kqffvppsfuysrLc35XSr18/GYZR4hdGud4h1a9fX40aNSr2bbSvv/566YP8L9c7p/PfbRUWFnr1WG8MHTpU27dv17Jly9SwYUOvD6dbrVYNGDBAH3/8sd577z0VFRUVO3Xjejd2fu179uzR119/XeE6e/bsqeDgYK1cudJjfSV9pNRmsxV7d7pp06Zi1xS4PhFxYViMjo5W69at9corr3hc6+HiOnVns9l03XXXaevWrR4B88CBA/rss8+8HpvrD926dev06aefenwSR6q+fcCX8+NS2jY9deqUx22r1er+uLA3R5y2bt3qMX979+7Vnj173PvrJZdcou7du2vNmjU6fvx4sce75qwyLqw9JCRE7du3l2EY7lNeJR15WLFihVdHQyujvO1Rkn79+slms2nJkiXFnh+GYRQbJzhSggq49dZby23Tt29fxcXF6amnnlJaWpqioqK0fft2ffjhhxo/fnyxw7nnmzRpkj766CNNmzZNt956q6666irl5+frxx9/1ObNm/Xhhx8qIiJC1157rYYOHaoVK1bo0KFDuv766+V0OvXPf/5TcXFxGjNmjCTptttu04svvqgHHnhA0dHR+vLLL716N92lSxeFh4dr1qxZGjt2rCwWi959912vDpl7Y/Dgwfr73/+uLVu26I477lBwcLDXj01MTNSKFSu0aNEiRUZGuq/VcenTp48++OADJSUlqU+fPjpy5IhWr16tyy+/vMJfVhUREaGJEyfqhRde0NSpU9W7d29999132rZtm8fRD1e/zzzzjGbPnq0uXbroxx9/1Pr16z2OsEhyX4i4evVqhYWFKTQ0VDExMWrVqpXmzZunyZMna/DgwRo2bJiaNWum9PR07dy5U/Xr19fzzz8vSZo+fbo+/fRT/fa3v9Udd9whh8OhlStX6vLLL9e+ffu8GttVV12lNm3a6KmnnvL4SLVLde0Dvpyf88cinfvI63XXXSebzaZBgwbpz3/+szIzM3XttdeqWbNmOnr0qFauXKkOHToU229K0rp1a91xxx264447VFhYqOXLl6thw4Yep1cfeughjR49WkOGDNHtt9+uVq1a6ddff9XXX3+tY8eO6b333qvUmCZNmqTGjRura9euuuSSS5SamqqVK1eqd+/e7iOwffr00bvvvqv69evr8ssv19dff60dO3ZU27Vm3myPkh7zhz/8QU888YTS0tKUkJCgsLAwHTlyRFu3btXtt9+uSZMmVUu9gYpQAp+yWq167rnntGjRIm3cuFFr165VixYtdN9992nixIllPrZevXpasWKFXnjhBSUnJ+udd95R/fr11bZtW02fPt3juovHHntMUVFReuutt7RgwQI1aNBA0dHR6tKli7tNUlKSMjIytHnzZm3atEnx8fF6+eWXS72Q1qVRo0Z6/vnn9be//U1PP/207Ha7br75ZvXo0cMnLyCNGzdWr169lJKSoqFDh1bosV27dtWll16q//znP8X+kErSsGHD3N8R8tlnn+nyyy/X3//+dyUnJ2vXrl0VrvUPf/iDQkJCtHr1au3cuVMxMTF65ZVXiv3e0bRp05Sfn6/169dr48aN6tixo1544QU98cQTHu2Cg4M1f/58Pfnkk3r44YdVVFSkxx57TK1atVJcXJzWrFmjZ599VitXrlReXp6aNGmimJgYjRw50r2OK6+8UkuXLtVjjz2mRYsWqXnz5po+fbpOnDjhdSiRzgW8559/Xm3atHH/YXeprn3A1/MjnXs3PnbsWL3//vt67733ZBiG+/tS3njjDb3++uvKyspSkyZNlJiYqOnTp3t1HcUtt9wiq9WqZcuW6eTJk4qJidGDDz6opk2buttcfvnlevvtt7VkyRKtW7dOp0+fVkREhDp27KikpKRKjUeSRo4cqfXr1+vVV19VXl6emjdvrrFjx+r3v/+9u80DDzwgq9Wq9evXq6CgQF27dtWrr75aZkioCm+2R0mmTJmitm3b6rXXXtMzzzwjSWrevLl69epV5lHn2spi+OrtHwCvJSUl6ccff9SWLVv8XQrg4ciRI7rxxht133338S5ebI+LjWtKgIvs+PHjlTpKAgA1HadvgIvk8OHD+te//qW33npLQUFBHqckAAAcKQEumt27d+u+++7TkSNHNH/+fH5hGQAuwDUlAADAFDhSAgAATIFQAgAATIFQAgAATIFQAgAATIFQAgAATIFQAgAATIFQAgAATIFQAgAATOH/A17/Wo84Cy8sAAAAAElFTkSuQmCC\n"
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Generate a base64 encoding of variant identifier fields in order to better match variants across samples.\n",
        "def get_var_base64_encoding_row(row):\n",
        "  return get_var_base64_encoding(row[\"chr\"], row[\"hg38_pos\"], row[\"ref\"], (row[\"alt\"],))\n",
        "\n",
        "def get_var_base64_encoding(chrom: str, pos: int, ref: str, alts: Tuple[str, ...]) -> str:\n",
        "  id_phrase = \"_\".join([chrom.lower().replace(\"chr\", \"\"), str(pos), ref, *[x for x in alts]])\n",
        "  b64encode = base64.b64encode(id_phrase.encode(\"utf-8\")).decode(\"utf-8\")\n",
        "  return b64encode"
      ],
      "metadata": {
        "id": "I5R14X2K9l5-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "b64encode_col = \"b64encode\"\n",
        "mol_val_subset_df = molecularly_val_df.loc[:, [\"chr\", \"hg38_pos\", \"ref\", \"alt\", \"ANY_PUB\", \"WES_vaf\", \"WGS_id\"]].copy()\n",
        "mol_val_subset_df[b64encode_col] = mol_val_subset_df.apply(get_var_base64_encoding_row, axis=1)"
      ],
      "metadata": {
        "id": "XnEl_qVq8y9i"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "mol_val_subset_df.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "C-WuIRCo_KdL",
        "outputId": "c2dcec5f-4ab8-4edc-daa3-b85ade307840"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  chr   hg38_pos ref alt   ANY_PUB   WES_vaf    WGS_id  \\\n",
              "0  13  101237830   G   A  germline  0.306667  SSC02254   \n",
              "1   7   77942799   G   A    mosaic  0.076087  SSC02497   \n",
              "2  12   56254338   C   A    mosaic  0.327869  SSC02518   \n",
              "3  14   65629547   C   T    mosaic  0.055556  SSC00936   \n",
              "4  16   16136537   G   A    mosaic  0.193548  SSC01066   \n",
              "\n",
              "                  b64encode  \n",
              "0  MTNfMTAxMjM3ODMwX0dfQQ==  \n",
              "1      N183Nzk0Mjc5OV9HX0E=  \n",
              "2      MTJfNTYyNTQzMzhfQ19B  \n",
              "3      MTRfNjU2Mjk1NDdfQ19U  \n",
              "4      MTZfMTYxMzY1MzdfR19B  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-9496419f-b823-4973-a2d5-0f12c537035f\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chr</th>\n",
              "      <th>hg38_pos</th>\n",
              "      <th>ref</th>\n",
              "      <th>alt</th>\n",
              "      <th>ANY_PUB</th>\n",
              "      <th>WES_vaf</th>\n",
              "      <th>WGS_id</th>\n",
              "      <th>b64encode</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>13</td>\n",
              "      <td>101237830</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>germline</td>\n",
              "      <td>0.306667</td>\n",
              "      <td>SSC02254</td>\n",
              "      <td>MTNfMTAxMjM3ODMwX0dfQQ==</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>7</td>\n",
              "      <td>77942799</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>0.076087</td>\n",
              "      <td>SSC02497</td>\n",
              "      <td>N183Nzk0Mjc5OV9HX0E=</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>12</td>\n",
              "      <td>56254338</td>\n",
              "      <td>C</td>\n",
              "      <td>A</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>0.327869</td>\n",
              "      <td>SSC02518</td>\n",
              "      <td>MTJfNTYyNTQzMzhfQ19B</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>14</td>\n",
              "      <td>65629547</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>0.055556</td>\n",
              "      <td>SSC00936</td>\n",
              "      <td>MTRfNjU2Mjk1NDdfQ19U</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>16</td>\n",
              "      <td>16136537</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>0.193548</td>\n",
              "      <td>SSC01066</td>\n",
              "      <td>MTZfMTYxMzY1MzdfR19B</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-9496419f-b823-4973-a2d5-0f12c537035f')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  \n",
              "    <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-9496419f-b823-4973-a2d5-0f12c537035f')\"\n",
              "            title=\"Generate charts.\"\n",
              "            style=\"display:none;\">\n",
              "      \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "      <g>\n",
              "          <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "      </g>\n",
              "  </svg>\n",
              "    </button>\n",
              "   \n",
              "  <style>\n",
              "    .colab-df-quickchart {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-quickchart:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-quickchart {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-quickchart:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#df-9496419f-b823-4973-a2d5-0f12c537035f button.colab-df-quickchart');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function quickchart(key) {\n",
              "        const containerElement = document.querySelector('#df-9496419f-b823-4973-a2d5-0f12c537035f');\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'generateCharts', [key], {})\n",
              "      }\n",
              "    </script>    \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-9496419f-b823-4973-a2d5-0f12c537035f button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-9496419f-b823-4973-a2d5-0f12c537035f');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 22
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def pzm_or_not(x):\n",
        "  # This will happen if the dataframe was already converted.\n",
        "  if isinstance(x, bool):\n",
        "    return x\n",
        "  if x.lower() == \"mosaic\":\n",
        "    return True\n",
        "  if x.lower() in [\"germline\", \"false_positive\"]:\n",
        "    return False\n",
        "  else:\n",
        "    return np.nan\n",
        "\n",
        "# mol_val_subset_df.rename(columns={\"ANY_PUB\": \"is_pzm\"}, inplace=True)\n",
        "mol_val_subset_df[\"is_pzm\"] = mol_val_subset_df[\"ANY_PUB\"].apply(pzm_or_not)\n",
        "\n",
        "print(f\"\\n{mol_val_subset_df.shape}\")\n",
        "mol_val_subset_df.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 241
        },
        "id": "CIV1Njf9BVxD",
        "outputId": "23763f5d-8c9c-4ea5-fc2b-2b97f47b61b5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "(1635, 9)\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  chr   hg38_pos ref alt   ANY_PUB   WES_vaf    WGS_id  \\\n",
              "0  13  101237830   G   A  germline  0.306667  SSC02254   \n",
              "1   7   77942799   G   A    mosaic  0.076087  SSC02497   \n",
              "2  12   56254338   C   A    mosaic  0.327869  SSC02518   \n",
              "3  14   65629547   C   T    mosaic  0.055556  SSC00936   \n",
              "4  16   16136537   G   A    mosaic  0.193548  SSC01066   \n",
              "\n",
              "                  b64encode is_pzm  \n",
              "0  MTNfMTAxMjM3ODMwX0dfQQ==  False  \n",
              "1      N183Nzk0Mjc5OV9HX0E=   True  \n",
              "2      MTJfNTYyNTQzMzhfQ19B   True  \n",
              "3      MTRfNjU2Mjk1NDdfQ19U   True  \n",
              "4      MTZfMTYxMzY1MzdfR19B   True  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-2560e7e8-0545-4284-9a96-1f486761f47b\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chr</th>\n",
              "      <th>hg38_pos</th>\n",
              "      <th>ref</th>\n",
              "      <th>alt</th>\n",
              "      <th>ANY_PUB</th>\n",
              "      <th>WES_vaf</th>\n",
              "      <th>WGS_id</th>\n",
              "      <th>b64encode</th>\n",
              "      <th>is_pzm</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>13</td>\n",
              "      <td>101237830</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>germline</td>\n",
              "      <td>0.306667</td>\n",
              "      <td>SSC02254</td>\n",
              "      <td>MTNfMTAxMjM3ODMwX0dfQQ==</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>7</td>\n",
              "      <td>77942799</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>0.076087</td>\n",
              "      <td>SSC02497</td>\n",
              "      <td>N183Nzk0Mjc5OV9HX0E=</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>12</td>\n",
              "      <td>56254338</td>\n",
              "      <td>C</td>\n",
              "      <td>A</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>0.327869</td>\n",
              "      <td>SSC02518</td>\n",
              "      <td>MTJfNTYyNTQzMzhfQ19B</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>14</td>\n",
              "      <td>65629547</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>0.055556</td>\n",
              "      <td>SSC00936</td>\n",
              "      <td>MTRfNjU2Mjk1NDdfQ19U</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>16</td>\n",
              "      <td>16136537</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>0.193548</td>\n",
              "      <td>SSC01066</td>\n",
              "      <td>MTZfMTYxMzY1MzdfR19B</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-2560e7e8-0545-4284-9a96-1f486761f47b')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  \n",
              "    <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-2560e7e8-0545-4284-9a96-1f486761f47b')\"\n",
              "            title=\"Generate charts.\"\n",
              "            style=\"display:none;\">\n",
              "      \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "      <g>\n",
              "          <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "      </g>\n",
              "  </svg>\n",
              "    </button>\n",
              "   \n",
              "  <style>\n",
              "    .colab-df-quickchart {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-quickchart:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-quickchart {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-quickchart:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#df-2560e7e8-0545-4284-9a96-1f486761f47b button.colab-df-quickchart');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function quickchart(key) {\n",
              "        const containerElement = document.querySelector('#df-2560e7e8-0545-4284-9a96-1f486761f47b');\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'generateCharts', [key], {})\n",
              "      }\n",
              "    </script>    \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-2560e7e8-0545-4284-9a96-1f486761f47b button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-2560e7e8-0545-4284-9a96-1f486761f47b');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 23
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "mol_val_subset_df[mol_val_subset_df[\"hg38_pos\"] == 14692164]\n",
        "\n",
        "# MV8xNDY5MjE2NF9DX1Q=\n",
        "# ----------------\n",
        "\n",
        "# ------- appending\n",
        "# {'chrom': 'chr1', 'pos': 14692164, 'ref': 'C', 'alts': ('T',), 'b64encode': 'MV8xNDY5MjE2NF9DX1Q=',"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 81
        },
        "id": "TtS2k0H_dO1d",
        "outputId": "ab1da47e-c654-4859-c81d-e482be5595f4"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "    chr  hg38_pos ref alt         ANY_PUB   WES_vaf    WGS_id  \\\n",
              "116   1  14692164   T   C  false_positive  0.076923  SSC03604   \n",
              "\n",
              "                b64encode is_pzm  \n",
              "116  MV8xNDY5MjE2NF9UX0M=  False  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-4e2606f4-20a4-41b7-8ba4-88720902be70\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chr</th>\n",
              "      <th>hg38_pos</th>\n",
              "      <th>ref</th>\n",
              "      <th>alt</th>\n",
              "      <th>ANY_PUB</th>\n",
              "      <th>WES_vaf</th>\n",
              "      <th>WGS_id</th>\n",
              "      <th>b64encode</th>\n",
              "      <th>is_pzm</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>116</th>\n",
              "      <td>1</td>\n",
              "      <td>14692164</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>false_positive</td>\n",
              "      <td>0.076923</td>\n",
              "      <td>SSC03604</td>\n",
              "      <td>MV8xNDY5MjE2NF9UX0M=</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-4e2606f4-20a4-41b7-8ba4-88720902be70')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  \n",
              "    <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-4e2606f4-20a4-41b7-8ba4-88720902be70')\"\n",
              "            title=\"Generate charts.\"\n",
              "            style=\"display:none;\">\n",
              "      \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "      <g>\n",
              "          <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "      </g>\n",
              "  </svg>\n",
              "    </button>\n",
              "   \n",
              "  <style>\n",
              "    .colab-df-quickchart {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-quickchart:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-quickchart {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-quickchart:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#df-4e2606f4-20a4-41b7-8ba4-88720902be70 button.colab-df-quickchart');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function quickchart(key) {\n",
              "        const containerElement = document.querySelector('#df-4e2606f4-20a4-41b7-8ba4-88720902be70');\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'generateCharts', [key], {})\n",
              "      }\n",
              "    </script>    \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-4e2606f4-20a4-41b7-8ba4-88720902be70 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-4e2606f4-20a4-41b7-8ba4-88720902be70');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 24
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Remove molecularly validated sites with no consensus between publications"
      ],
      "metadata": {
        "id": "ZORHy5SaHvYe"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "conflict_count = mol_val_subset_df[\"is_pzm\"].isna().sum()\n",
        "if conflict_count > 0:\n",
        "  print(f\"The DataFrame contains `{conflict_count}` variants with 'conflict' type (i.e., no consensus between publications). These variants will be all removed.\")\n",
        "  mol_val_subset_df_ = mol_val_subset_df.dropna(subset=[\"is_pzm\"])\n",
        "  assert len(mol_val_subset_df) - len(mol_val_subset_df_) == conflict_count, \"Failed removing variants with 'conflict' type.\"\n",
        "  mol_val_subset_df = mol_val_subset_df_\n",
        "\n",
        "print(f\"\\n{mol_val_subset_df.shape}\")\n",
        "mol_val_subset_df.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 258
        },
        "id": "UFk0fv6nbcD0",
        "outputId": "bd13f01b-9469-4a55-d912-1e7ae17c3784"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The DataFrame contains `2` variants with 'conflict' type (i.e., no consensus between publications). These variants will be all removed.\n",
            "\n",
            "(1633, 9)\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  chr   hg38_pos ref alt   ANY_PUB   WES_vaf    WGS_id  \\\n",
              "0  13  101237830   G   A  germline  0.306667  SSC02254   \n",
              "1   7   77942799   G   A    mosaic  0.076087  SSC02497   \n",
              "2  12   56254338   C   A    mosaic  0.327869  SSC02518   \n",
              "3  14   65629547   C   T    mosaic  0.055556  SSC00936   \n",
              "4  16   16136537   G   A    mosaic  0.193548  SSC01066   \n",
              "\n",
              "                  b64encode is_pzm  \n",
              "0  MTNfMTAxMjM3ODMwX0dfQQ==  False  \n",
              "1      N183Nzk0Mjc5OV9HX0E=   True  \n",
              "2      MTJfNTYyNTQzMzhfQ19B   True  \n",
              "3      MTRfNjU2Mjk1NDdfQ19U   True  \n",
              "4      MTZfMTYxMzY1MzdfR19B   True  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-8954a38e-9d54-43ed-843f-3991a3f9fa3e\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chr</th>\n",
              "      <th>hg38_pos</th>\n",
              "      <th>ref</th>\n",
              "      <th>alt</th>\n",
              "      <th>ANY_PUB</th>\n",
              "      <th>WES_vaf</th>\n",
              "      <th>WGS_id</th>\n",
              "      <th>b64encode</th>\n",
              "      <th>is_pzm</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>13</td>\n",
              "      <td>101237830</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>germline</td>\n",
              "      <td>0.306667</td>\n",
              "      <td>SSC02254</td>\n",
              "      <td>MTNfMTAxMjM3ODMwX0dfQQ==</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>7</td>\n",
              "      <td>77942799</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>0.076087</td>\n",
              "      <td>SSC02497</td>\n",
              "      <td>N183Nzk0Mjc5OV9HX0E=</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>12</td>\n",
              "      <td>56254338</td>\n",
              "      <td>C</td>\n",
              "      <td>A</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>0.327869</td>\n",
              "      <td>SSC02518</td>\n",
              "      <td>MTJfNTYyNTQzMzhfQ19B</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>14</td>\n",
              "      <td>65629547</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>0.055556</td>\n",
              "      <td>SSC00936</td>\n",
              "      <td>MTRfNjU2Mjk1NDdfQ19U</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>16</td>\n",
              "      <td>16136537</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>0.193548</td>\n",
              "      <td>SSC01066</td>\n",
              "      <td>MTZfMTYxMzY1MzdfR19B</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-8954a38e-9d54-43ed-843f-3991a3f9fa3e')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  \n",
              "    <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-8954a38e-9d54-43ed-843f-3991a3f9fa3e')\"\n",
              "            title=\"Generate charts.\"\n",
              "            style=\"display:none;\">\n",
              "      \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "      <g>\n",
              "          <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "      </g>\n",
              "  </svg>\n",
              "    </button>\n",
              "   \n",
              "  <style>\n",
              "    .colab-df-quickchart {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-quickchart:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-quickchart {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-quickchart:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#df-8954a38e-9d54-43ed-843f-3991a3f9fa3e button.colab-df-quickchart');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function quickchart(key) {\n",
              "        const containerElement = document.querySelector('#df-8954a38e-9d54-43ed-843f-3991a3f9fa3e');\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'generateCharts', [key], {})\n",
              "      }\n",
              "    </script>    \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-8954a38e-9d54-43ed-843f-3991a3f9fa3e button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-8954a38e-9d54-43ed-843f-3991a3f9fa3e');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 25
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "@dataclass\n",
        "class ValidatedVarInfo:\n",
        "  is_pzm: bool\n",
        "  any_pub: str\n",
        "  sample_id: str\n",
        "\n",
        "validated_var_dict = {}\n",
        "for _, r in mol_val_subset_df.iterrows():\n",
        "  key = r[b64encode_col]\n",
        "  value = ValidatedVarInfo(is_pzm=r[\"is_pzm\"], any_pub=r[\"ANY_PUB\"], sample_id=r[\"WGS_id\"])\n",
        "  validated_var_dict[key] = value\n",
        "\n",
        "# molecularly_val_dict = mol_val_subset_df.set_index(b64encode_col)[\"is_pzm\"].to_dict()\n",
        "# molecularly_val_any_pub_dict = mol_val_subset_df.set_index(b64encode_col)[\"ANY_PUB\"].to_dict()"
      ],
      "metadata": {
        "id": "-RXY7Dij_eyt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(\"b64 encoding\\tis pzm?\\tany pub?\\tsample id\")\n",
        "for k, v in list(validated_var_dict.items())[: 3]:\n",
        "  print(f\"{k}\\t{v.is_pzm}\\t{v.any_pub}\\t{v.sample_id}\")\n",
        "\n",
        "# print(\"\\nany pub?\")\n",
        "# for k, v in list(molecularly_val_any_pub_dict.items())[: 3]:\n",
        "#   print(f\"{k} : {v}\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MphG9Jy6_2gJ",
        "outputId": "db2e0dbf-acad-464f-fbca-b87de56b7a30"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "b64 encoding\tis pzm?\tany pub?\tsample id\n",
            "MTNfMTAxMjM3ODMwX0dfQQ==\tFalse\tgermline\tSSC02254\n",
            "N183Nzk0Mjc5OV9HX0E=\tTrue\tmosaic\tSSC02497\n",
            "MTJfNTYyNTQzMzhfQ19B\tTrue\tmosaic\tSSC02518\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Read VCF files into Pandas DataFrame"
      ],
      "metadata": {
        "id": "tGhNJI5uOGid"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def get_sor(sb):\n",
        "  ref_fw = sb[0] + 1\n",
        "  ref_rv = sb[1] + 1\n",
        "  alt_fw = sb[2] + 1\n",
        "  alt_rv = sb[3] + 1\n",
        "  symmetricalRatio = (ref_fw * alt_rv) / (alt_fw * ref_rv) + (alt_fw * ref_rv) / (ref_fw * alt_rv)\n",
        "  refRatio = ref_rv / ref_fw\n",
        "  altRatio = alt_fw / alt_rv\n",
        "  sor = math.log(symmetricalRatio) + math.log(refRatio) - math.log(altRatio)\n",
        "  return sor"
      ],
      "metadata": {
        "id": "Ie8CBA3GZ7xQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "no_repeat_masker_key = \"no_repeat_masker\"\n",
        "rmcl_values = [\"DNA\", \"DNA?\", \"LINE\", \"LTR\", \"LTR?\", \"Low_complexity\", \"RC\", \"RC?\", \"RMCL\", \"RNA\", \"Retroposon\", \"SINE\", \"SINE?\", \"Satellite\", \"Simple_repeat\", \"Unknown\", \"rRNA\", \"scRNA\", \"snRNA\", \"srpRNA\", \"tRNA\", no_repeat_masker_key]\n",
        "rmcl_dict = {rmcl_values[i]: i for i in range(len(rmcl_values))}\n",
        "\n",
        "def get_one_hot_rmcl_encoding(rmcl):\n",
        "  one_hot = np.zeros((len(rmcl_dict), ))\n",
        "  one_hot[rmcl_dict[rmcl]] = 1\n",
        "  return one_hot\n",
        "\n",
        "def get_rmcl_encoding(rmcl: Tuple[str, ...]):\n",
        "  one_hot_sequence = [get_one_hot_rmcl_encoding(x) for x in rmcl]\n",
        "  one_hot_array = np.array(one_hot_sequence)\n",
        "\n",
        "  # Get the index of the maximum value in each row\n",
        "  row_max_index = np.argmax(one_hot_array, axis=1)\n",
        "\n",
        "  # Convert the row indices to a single number\n",
        "  number = np.sum(row_max_index * (len(rmcl_dict) ** np.arange(len(row_max_index))))\n",
        "\n",
        "  return number\n",
        "\n",
        "# TEST\n",
        "print(get_rmcl_encoding((\"DNA\",)))\n",
        "print(get_rmcl_encoding((\"DNA\", \"DNA?\")))\n",
        "print(get_rmcl_encoding((\"DNA\", \"rRNA\", \"SINE\", \"RC\")))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0wMp0i-rIu3N",
        "outputId": "1e29acf8-f529-4699-a9ab-1a5f2375e24a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "0\n",
            "22\n",
            "69564\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "@dataclass\n",
        "class Mapper:\n",
        "  is_sample_metric: bool\n",
        "  default: dict\n",
        "  func: Callable[[Any], dict]"
      ],
      "metadata": {
        "id": "ITCarT97uSXK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# INFO fields (contains both per-sample and per-site metrics)\n",
        "info_fields = {\n",
        "    # Phred-scaled quality that alt alleles are not germline variants\n",
        "    \"GERMQ\": Mapper(True, {}, lambda x: {\"GERMQ\": int(x)}),\n",
        "\n",
        "    # median base quality by allele\n",
        "    \"MBQ\": Mapper(True, {}, lambda x: {\"MBQ_ref\": int(x[0]), \"MBQ_alt\": int(x[1])}),\n",
        "\n",
        "    # median fragment length by allele\n",
        "    \"MFRL\": Mapper(True, {}, lambda x: {\"MFRL_ref\": int(x[0]), \"MFRL_alt\": int(x[1])}),\n",
        "\n",
        "    # median mapping quality by allele\n",
        "    \"MMQ\": Mapper(True, {}, lambda x: {\"MMQ_ref\": int(x[0]), \"MMQ_alt\": int(x[1])}),\n",
        "\n",
        "    # median distance from end of read\n",
        "    \"MPOS\": Mapper(True, {}, lambda x: {\"MPOS\": int(x[0])}),\n",
        "\n",
        "    # Normalized, Phred-scaled likelihoods for genotypes as defined\n",
        "    # in the VCF specification\n",
        "    \"PL\": Mapper(True, {}, lambda x: {\"PL\": int(x)}),\n",
        "\n",
        "    # Log 10 likelihood ratio score of variant existing versus not existing\n",
        "    \"TLOD\": Mapper(True, {}, lambda x: {\"TLOD\": float(x[0])}),\n",
        "\n",
        "    \"all_cohort_af\": Mapper(False, {}, lambda x: {\"all_cohort_af\": float(x[0])}),\n",
        "\n",
        "    # Number of events in this haplotype\n",
        "    \"ECNT\": Mapper(False, {}, lambda x: {\"ECNT\": int(x)}),\n",
        "\n",
        "    # Genotype Quality\n",
        "    # The GC percent track shows the percentage of G (guanine)\n",
        "    # and C (cytosine) bases in 5-base windows. High GC content\n",
        "    # is typically associated with gene-rich areas.\n",
        "    \"GC\": Mapper(False, {}, lambda x: {\"GC\": float(max(x))}),\n",
        "\n",
        "    # If present, site occurs in a low complexity region\n",
        "    \"LCR\": Mapper(False, {\"LCR\": False}, lambda x: {\"LCR\": True}),\n",
        "\n",
        "    # Frequency of PASS alleles in the cohort run with the same PON\n",
        "    \"pass_cohort_AF\": Mapper(False, {}, lambda x: {\"pass_cohort_AF\": float(x)}),\n",
        "\n",
        "    # negative log 10 population allele frequencies of alt alleles\n",
        "    \"POPAF\": Mapper(False, {}, lambda x: {\"POPAF\": float(x[0])}),\n",
        "\n",
        "    # Repeat Masker class\n",
        "    \"RMCL\": Mapper(False, {\"RMCL\": get_rmcl_encoding((no_repeat_masker_key,))}, lambda x: {\"RMCL\": get_rmcl_encoding(x)}),\n",
        "\n",
        "    # Number of times tandem repeat unit is repeated,\n",
        "    # for each allele (including reference)\n",
        "    \"RPA\": Mapper(False, {}, lambda x: {\"RPA_ref\": int(x[0]), \"RPA_alt\": int(x[1])}),\n",
        "\n",
        "    # If present, site occurs in a segmental duplication region\n",
        "    \"SEGDUP\": Mapper(False, {\"SEGDUP\": False}, lambda x: {\"SEGDUP\": True}),\n",
        "\n",
        "    # If present, site occurs in a simple repeat region\n",
        "    \"SIMPLEREP\": Mapper(False, {\"SIMPLEREP\": False}, lambda x: {\"SIMPLEREP\": True}),\n",
        "\n",
        "    # Variant is a short tandem repeat\n",
        "    \"STR\": Mapper(False, {\"STR\": False}, lambda x: {\"STR\": True}),\n",
        "\n",
        "    # Phred-scaled quality that alt alleles in STRs are not\n",
        "    # polymerase slippage errors\n",
        "    \"STRQ\": Mapper(False, {}, lambda x: {\"STRQ\": int(x)}),\n",
        "\n",
        "    # If present, site occurs in the broad exome evaluation region\n",
        "    \"WESREG\": Mapper(False, {\"WESREG\": False}, lambda x: {\"WESREG\": True})\n",
        "  }"
      ],
      "metadata": {
        "id": "y9mF2S7qvSbG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# FORMAT fields (contains both per-sample and per-site metrics)\n",
        "genotype_fields = {\n",
        "\n",
        "    # Allelic depths for the ref and alt alleles in the order listed\n",
        "    \"AD\": Mapper(True, {}, lambda x: {\"AD_ref\": int(x[0]), \"AD_alt\": int(x[1])}),\n",
        "\n",
        "    # Allele fractions of alternate alleles in the tumor\n",
        "    \"AF\": Mapper(True, {}, lambda x: {\"AF\": float(x[0])}),\n",
        "\n",
        "    # Approximate read depth (reads with MQ=255 or with bad mates are filtered)\n",
        "    \"DP\": Mapper(True, {}, lambda x: {\"DP\": int(x)}),\n",
        "\n",
        "    # Count of fragments supporting each allele.\n",
        "    \"FAD\": Mapper(True, {}, lambda x: {\"FAD_ref\": int(x[0]), \"FAD_alt\": int(x[1])}),\n",
        "\n",
        "    # Genotype Quality\n",
        "    \"GQ\": Mapper(True, {}, lambda x: {\"GQ\": int(x)}),\n",
        "\n",
        "    # 0: phased (e.g., 1|0 or 0|1)\n",
        "    # 1: unphased (e.g., 0/1)\n",
        "    \"GT\": Mapper(True, {}, lambda x: {\"GT\": 0 if \"|\" in x else 1}),\n",
        "\n",
        "    # \"Per-sample component statistics which comprise the Fisher's\n",
        "    # exact test to detect strand bias.\"\n",
        "    \"SB\": Mapper(True, {}, lambda x: {\"SOR\": get_sor(x)}),\n",
        "\n",
        "    # # Count of reads in F1R2 pair orientation supporting each allele\n",
        "    # \"F1R2\": lambda x: {\"F1R2_allele_a\": float(x[0]), \"F1R2_allele_b\": float(x[1])},\n",
        "\n",
        "    # # Count of reads in F2R1 pair orientation supporting each allele\n",
        "    # \"F2R1\": lambda x: {\"F2R1_allele_a\": float(x[0]), \"F2R1_allele_b\": float(x[1])},\n",
        "  }"
      ],
      "metadata": {
        "id": "BoNbph-uvb2g"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Asserts if the given VCF file contains the defined info and genotype fields.\n",
        "def assert_fields(vcf_filename):\n",
        "  vcf = VariantFile(vcf_filename)\n",
        "  invalid_info_fields = list(set(info_fields.keys()) - set(vcf.header.info.keys()))\n",
        "  if invalid_info_fields:\n",
        "    raise ValueError(\n",
        "        f\"The VCF file does not contain the following info \" +\n",
        "        f\"fields in the given list: {invalid_info_fields}\")"
      ],
      "metadata": {
        "id": "9YZRih0mE7NZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# DEBUG\n",
        "# def print_variant(variant, mvv):\n",
        "#   if variant.pos == TEST_VAR_POS:\n",
        "#     print(\"\\n-------------- parsing var\")\n",
        "#     print(variant)\n",
        "#     print(\"\\n\")\n",
        "#     print(mvv)\n",
        "#     print(\"----------------\\n\")"
      ],
      "metadata": {
        "id": "8wZIYPaVYhTQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def parse_variant(variant, mappers, sample_id):\n",
        "  b64encode = get_var_base64_encoding(variant.chrom, variant.pos, variant.ref, variant.alts)\n",
        "\n",
        "  # mvv: molecularly validated variant\n",
        "  mvv = validated_var_dict.get(b64encode)\n",
        "\n",
        "  # DEBUG\n",
        "  # print_variant(variant, mvv)\n",
        "  # if variant.pos == 104950377:\n",
        "  #   print(\"\\n-------------- parsing var ---- 2\")\n",
        "  #   print(mvv)\n",
        "  #   print(b64encode)\n",
        "  #   print(\"----------------\\n\")\n",
        "\n",
        "  # Fields that should be included in both sample and site level dataframes.\n",
        "  common_fields = {\n",
        "    \"chrom\": variant.chrom,\n",
        "    \"pos\": variant.pos,\n",
        "    \"ref\": variant.ref,\n",
        "    \"alts\": variant.alts,\n",
        "    b64encode_col: b64encode,\n",
        "    \"filter\": \";\".join(x for x in variant.filter.keys()),\n",
        "    \"any_pub\": mvv.any_pub if mvv is not None else np.nan\n",
        "  }\n",
        "\n",
        "  sample_metrics = {}\n",
        "  site_metrics = {}\n",
        "  for k, v in common_fields.items():\n",
        "    sample_metrics[k] = v\n",
        "    site_metrics[k] = v\n",
        "\n",
        "  is_validated_pzm_key = \"is_molecularly_validated_pzm\"\n",
        "  sample_metrics[is_validated_pzm_key] = mvv.is_pzm if mvv is not None and mvv.sample_id == sample_id else np.nan\n",
        "  # site_metrics[is_validated_pzm_key] = mvv.is_pzm if mvv is not None else np.nan\n",
        "  site_metrics[is_validated_pzm_key] = mvv.is_pzm if mvv is not None and mvv.sample_id == sample_id else np.nan\n",
        "\n",
        "  # # DEBUG\n",
        "  # xx = mvv.is_pzm if mvv is not None and mvv.sample_id == sample_id else np.nan\n",
        "  # yy = mvv.is_pzm if mvv is not None else np.nan\n",
        "  # if not np.isnan(yy) and xx != yy:\n",
        "  #   print(\"---------------------------------------------------------- diff\")\n",
        "  #   print(xx)\n",
        "  #   print(yy)\n",
        "  #   print(variant)\n",
        "  #   print(\"^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n\\n\")\n",
        "\n",
        "  for k, mapper in mappers.items():\n",
        "    if k in info_fields and k in variant.info:\n",
        "      x = mapper.func(variant.info.get(k))\n",
        "    elif k in genotype_fields and k in variant.samples[0]:\n",
        "      x = mapper.func(variant.samples[0][k])\n",
        "    else:\n",
        "      x = mapper.default\n",
        "\n",
        "    if mapper.is_sample_metric:\n",
        "      sample_metrics.update(x)\n",
        "    else:\n",
        "      site_metrics.update(x)\n",
        "\n",
        "\n",
        "  # print(\"----------------\")\n",
        "  # print(dir(variant))\n",
        "  # print(variant.format.keys())\n",
        "  # print(\"info:\")\n",
        "  # print(variant.info.keys())\n",
        "  # print(dir(variant.samples[0]))\n",
        "  # print(variant.samples[0].keys())\n",
        "  # print(\"\\n\\n\")\n",
        "  # print(variant.samples[0][\"GQ\"])\n",
        "  # print(\"----------------\")\n",
        "  #sample_metrics[\"GQ\"] =   Mapper(True, {}, lambda x: {\"GQ\": int(x)}),\n",
        "\n",
        "  # True: the variant is SNV\n",
        "  # False: the variant is Indel\n",
        "  site_metrics[\"is_snv\"] = True if len(variant.ref) == len(variant.alts) else False\n",
        "\n",
        "  return sample_metrics, site_metrics"
      ],
      "metadata": {
        "id": "WX-zjA817sij"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "black_list_regions = [\n",
        "    [\"chr14\", 105586437, 106879844],\n",
        "    [\"chr2\", 88857361, 90235368],\n",
        "    [\"chr22\", 22026076, 22922913],\n",
        "    [\"chr6\", 28510120, 33480577]]"
      ],
      "metadata": {
        "id": "DxIiCCJFmy3j"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Parse the VCF\n",
        "def get_variants_df(sample_id, vcf_filename, unfiltered_vcf_filename):\n",
        "  vcf = VariantFile(vcf_filename)\n",
        "  parsed_vars = []\n",
        "  parsed_sample_metrics = []\n",
        "  parsed_site_metrics = []\n",
        "\n",
        "  mappers = {**info_fields, **genotype_fields}\n",
        "\n",
        "  # Get variants from the filtered mutect VCF, which, for instance, contains only the variants with Filter==PASS.\n",
        "  for variant in vcf:\n",
        "    # if variant.pos == 104950377:\n",
        "    #   print(\"--------------------------------***********- yes\")\n",
        "    sample_metrics, site_metrics = parse_variant(variant, mappers, sample_id=sample_id)\n",
        "\n",
        "    # if sample_metrics[\"AF\"] >= 0.3675:\n",
        "    #   continue\n",
        "\n",
        "    # if site_metrics[\"SIMPLEREP\"]:\n",
        "    #   continue\n",
        "\n",
        "    # if site_metrics[\"SEGDUP\"]:\n",
        "    #   continue\n",
        "\n",
        "    # site_metrics.head()\n",
        "    # break\n",
        "\n",
        "    # for region in black_list_regions:\n",
        "    #   if sample_metrics\n",
        "\n",
        "\n",
        "\n",
        "    print(site_metrics.keys())\n",
        "    print(site_metrics[\"SIMPLEREP\"])\n",
        "\n",
        "    parsed_sample_metrics.append(sample_metrics)\n",
        "    parsed_site_metrics.append(site_metrics)\n",
        "\n",
        "  # Get molecularly validated variants whose mutect2 Filter!=PASS.\n",
        "  # First, from the molecularly validated sites, get only those that belong to the given sample.\n",
        "  mol_val_vars = mol_val_subset_df[mol_val_subset_df[\"WGS_id\"] == sample_id]\n",
        "\n",
        "  # Second, iterate through ever one of those variants, fetch their corresponding\n",
        "  # variant from the unfiltered VCF, parse it, and add it to the list of the\n",
        "  # previously parsed variants as needed.\n",
        "  for _, row in mol_val_vars.iterrows():\n",
        "    unfiltered_vcf = VariantFile(unfiltered_vcf_filename)\n",
        "    iterator = unfiltered_vcf.fetch(\"chr\" + row[\"chr\"], row[\"hg38_pos\"]-1, row[\"hg38_pos\"])\n",
        "\n",
        "    # if row[\"hg38_pos\"] == TEST_VAR_POS:\n",
        "    #   print(\"----------- yes\")\n",
        "    #   print(row)\n",
        "    #   print(\"^^^\")\n",
        "\n",
        "    # print(f\"-------------------- {row['chr']}, {row['hg38_pos']}\")\n",
        "    # print(row)\n",
        "\n",
        "    for unfiltered_var in iterator:\n",
        "      # print(\"***************\")\n",
        "      if unfiltered_var.filter.keys()[0] == \"PASS\":\n",
        "        continue\n",
        "\n",
        "      # print(\"+++++++++++++\")\n",
        "\n",
        "      uf_sample_metrics, uf_site_metrics = parse_variant(unfiltered_var, mappers = {**info_fields, **genotype_fields}, sample_id=sample_id)\n",
        "      b64encode = get_var_base64_encoding(unfiltered_var.chrom, unfiltered_var.pos, unfiltered_var.ref, unfiltered_var.alts)\n",
        "      if uf_sample_metrics[\"b64encode\"] != b64encode:\n",
        "        print(\"Warning: skipping the unfiltered variant as its base64 encode does not match the expected.\")\n",
        "        continue\n",
        "\n",
        "      # print(\"------- appending\")\n",
        "      # print(uf_sample_metrics)\n",
        "      # print(\"\\n\\n\")\n",
        "      # print(uf_site_metrics)\n",
        "      # print(\"\\n\\n\")\n",
        "      parsed_sample_metrics.append(uf_sample_metrics)\n",
        "      parsed_site_metrics.append(uf_site_metrics)\n",
        "\n",
        "  sample_metrics_df = pd.DataFrame(parsed_sample_metrics)\n",
        "  site_metrics_df = pd.DataFrame(parsed_site_metrics)\n",
        "\n",
        "  sample_metrics_df = sample_metrics_df.rename(columns={x: x.lower() for x in sample_metrics_df.columns})\n",
        "  site_metrics_df = site_metrics_df.rename(columns={x: x.lower() for x in site_metrics_df.columns})\n",
        "\n",
        "  return sample_metrics_df, site_metrics_df"
      ],
      "metadata": {
        "id": "7lguN6VLJev1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# TEST\n",
        "if run_test:\n",
        "  _test_sample_id=\"SSC02602\"\n",
        "  TEST_VAR_POS = 11414455\n",
        "  _test_vcf_filename, _test_unfiltered_vcf_filename = localize_obj_if_needed(_test_sample_id)\n",
        "  print(_test_vcf_filename)\n",
        "  _test_sample_metrics_df, _test_site_metrics_df = get_variants_df(_test_sample_id, _test_vcf_filename, _test_unfiltered_vcf_filename)"
      ],
      "metadata": {
        "id": "pPtfnE7Ml1Y-",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 340
        },
        "outputId": "043ce3c2-42b2-4eed-b45c-c7cd34d98447"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "./pzm/SSC02602.vcf.gz\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "AttributeError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-56-958d7ca8b09a>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      5\u001b[0m   \u001b[0m_test_vcf_filename\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_test_unfiltered_vcf_filename\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlocalize_obj_if_needed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_test_sample_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m   \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_test_vcf_filename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m   \u001b[0m_test_sample_metrics_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_test_site_metrics_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_variants_df\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_test_sample_id\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_test_vcf_filename\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_test_unfiltered_vcf_filename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;32m<ipython-input-55-af7e9c92170f>\u001b[0m in \u001b[0;36mget_variants_df\u001b[0;34m(sample_id, vcf_filename, unfiltered_vcf_filename)\u001b[0m\n\u001b[1;32m     23\u001b[0m       \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m     \u001b[0msite_metrics\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     26\u001b[0m     \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mAttributeError\u001b[0m: 'dict' object has no attribute 'head'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# _test_sample_metrics_df[_test_sample_metrics_df[\"pos\"] == 104950377]"
      ],
      "metadata": {
        "id": "_cD4cKh9UTnk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Print summary stats.\n",
        "def print_summary_stats(sample_metrics_df, site_metrics_df):\n",
        "  # def print_df_stats(df_):\n",
        "  #   print(f\"\\t\\t* Parsed {len(df_):,} (shape: {df_.shape}\")\n",
        "  # def print_counts(df_):\n",
        "    # counts_dict = df_[\"is_molecularly_validated_pzm\"].value_counts(dropna=False).to_dict()\n",
        "    # print(\"\\t\\t* Overlapping with the molecularly validates variants (True=PZM; False=Not-PZM; nan=Unknown):\")\n",
        "    # print(f\"\\t\\t  {counts_dict}\")\n",
        "\n",
        "  def get_counts_df(df_):\n",
        "    dfs_ = []\n",
        "    for c in [\"any_pub\", \"filter\", \"is_molecularly_validated_pzm\"]:\n",
        "      x = df_[c].value_counts(dropna=False).to_frame().reset_index().rename(columns={\"index\": \"value\", c: \"count\"})\n",
        "      x[\"category\"] = c\n",
        "      dfs_.append(x)\n",
        "    return pd.concat(dfs_)\n",
        "\n",
        "  # print(\"\\t- Sample metrics:\")\n",
        "  # print_df_stats(sample_metrics_df)\n",
        "  # print_counts(sample_metrics_df)\n",
        "\n",
        "  # print(\"\\t- Site metrics:\")\n",
        "  # print_df_stats(site_metrics_df)\n",
        "  # print_counts(site_metrics_df)\n",
        "\n",
        "  sample_counts_df = get_counts_df(sample_metrics_df)\n",
        "  sample_counts_df[\"site_or_sample_level\"] = \"sample\"\n",
        "\n",
        "  site_counts_df = get_counts_df(site_metrics_df)\n",
        "  site_counts_df[\"site_or_sample_level\"] = \"site\"\n",
        "\n",
        "  all_counts = pd.concat([sample_counts_df, site_counts_df])\n",
        "  return all_counts[[\"site_or_sample_level\", \"category\", \"value\", \"count\"]]"
      ],
      "metadata": {
        "id": "ywStRKSrgOEx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# TEST\n",
        "if run_test:\n",
        "  _test_counts_df = print_summary_stats(_test_sample_metrics_df, _test_site_metrics_df)\n",
        "  print(_test_counts_df)"
      ],
      "metadata": {
        "id": "gAwwmqcOdg8s",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "a5e4aa45-8d2e-4d66-d71e-3808f013aa27"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  site_or_sample_level                      category  \\\n",
            "0               sample                       any_pub   \n",
            "1               sample                       any_pub   \n",
            "0               sample                        filter   \n",
            "1               sample                        filter   \n",
            "2               sample                        filter   \n",
            "0               sample  is_molecularly_validated_pzm   \n",
            "1               sample  is_molecularly_validated_pzm   \n",
            "0                 site                       any_pub   \n",
            "1                 site                       any_pub   \n",
            "0                 site                        filter   \n",
            "1                 site                        filter   \n",
            "2                 site                        filter   \n",
            "0                 site  is_molecularly_validated_pzm   \n",
            "1                 site  is_molecularly_validated_pzm   \n",
            "\n",
            "                                 value  count  \n",
            "0                                  NaN  11727  \n",
            "1                             germline      1  \n",
            "0                                 PASS  11726  \n",
            "1                             germline      1  \n",
            "2  clustered_events;germline;haplotype      1  \n",
            "0                                  NaN  11727  \n",
            "1                                False      1  \n",
            "0                                  NaN  11727  \n",
            "1                             germline      1  \n",
            "0                                 PASS  11726  \n",
            "1                             germline      1  \n",
            "2  clustered_events;germline;haplotype      1  \n",
            "0                                  NaN  11727  \n",
            "1                                False      1  \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# TEST\n",
        "if run_test:\n",
        "  print(_test_sample_metrics_df.shape)\n",
        "  _test_sample_metrics_df.head()"
      ],
      "metadata": {
        "id": "UQ-7JSEU0m0X",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "960e8343-155e-4cd5-95d3-e35c2683ac20"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "(11728, 25)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# TEST\n",
        "if run_test:\n",
        "  print(_test_site_metrics_df.shape)\n",
        "  _test_site_metrics_df.head()"
      ],
      "metadata": {
        "id": "7rmDG4NV0qqP",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "1b872b4d-e2fd-48f2-cdf4-bfff8bf9a39d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "(11728, 22)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def get_normalized_df(sample_id, vcf_filename, unfiltered_vcf_filename):\n",
        "  # assert_fields(vcf_filename)\n",
        "  sample_metrics_df, site_metrics_df = get_variants_df(sample_id, vcf_filename, unfiltered_vcf_filename)\n",
        "\n",
        "  sample_metrics_ndf = sample_metrics_df.copy()\n",
        "  site_metrics_ndf = site_metrics_df.copy()\n",
        "\n",
        "  no_norm_cols = [\"chrom\", \"pos\", \"ref\", \"alts\", \"b64encode\", \"filter\", \"any_pub\", \"is_molecularly_validated_pzm\"]\n",
        "\n",
        "  # std_scaler = StandardScaler() # Gaussian with zero mean and unit variance\n",
        "  mm_scaler = MinMaxScaler()\n",
        "  cols = np.setdiff1d(sample_metrics_ndf.columns.to_numpy(), no_norm_cols)\n",
        "  sample_metrics_ndf[cols] = mm_scaler.fit_transform(sample_metrics_ndf[cols])\n",
        "\n",
        "  # std_scaler = StandardScaler()\n",
        "  mm_scaler = MinMaxScaler()\n",
        "  cols = np.setdiff1d(site_metrics_ndf.columns.to_numpy(), no_norm_cols)\n",
        "  site_metrics_ndf[cols] = mm_scaler.fit_transform(site_metrics_ndf[cols])\n",
        "\n",
        "  return sample_metrics_df, sample_metrics_ndf, site_metrics_df, site_metrics_ndf"
      ],
      "metadata": {
        "id": "_iaquYQjFhof"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# TEST\n",
        "if run_test:\n",
        "  _test_sample_metrics_df, _test_sample_metrics_ndf, _test_site_metrics_df, _test_site_metrics_ndf = get_normalized_df(_test_sample_id, _test_vcf_filename, _test_unfiltered_vcf_filename)"
      ],
      "metadata": {
        "id": "7HZJ-f_dH8l5",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "f3a51d55-a550-44aa-e3ad-64f69a81eab5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "----------- yes\n",
            "chr                            16\n",
            "hg38_pos                 11414455\n",
            "ref                             C\n",
            "alt                             T\n",
            "ANY_PUB                  germline\n",
            "WES_vaf                  0.114754\n",
            "WGS_id                   SSC02602\n",
            "b64encode    MTZfMTE0MTQ0NTVfQ19U\n",
            "is_pzm                      False\n",
            "Name: 55, dtype: object\n",
            "^^^\n",
            "***************\n",
            "+++++++++++++\n",
            "\n",
            "-------------- parsing var\n",
            "chr16\t11414455\t.\tT\tC\t.\tgermline\tAS_FilterStatus=SITE;AS_SB_TABLE=12,3|9,8;DP=32;ECNT=1;GERMQ=1;MBQ=30,30;MFRL=392,402;MMQ=60,60;MPOS=31;POPAF=0.111;TLOD=47.83;WESREG;all_cohort_af=0.462327;cohort_n=1951\tGT:AD:AF:DP:F1R2:F2R1:FAD:SB\t0/1:15,17:0.529:32:11,10:4,7:15,17:12,3,9,8\n",
            "\n",
            "\n",
            "\n",
            "None\n",
            "----------------\n",
            "\n",
            "------- appending\n",
            "{'chrom': 'chr16', 'pos': 11414455, 'ref': 'T', 'alts': ('C',), 'b64encode': 'MTZfMTE0MTQ0NTVfVF9D', 'filter': 'germline', 'any_pub': nan, 'is_molecularly_validated_pzm': nan, 'GERMQ': 1, 'MBQ_ref': 30, 'MBQ_alt': 30, 'MFRL_ref': 392, 'MFRL_alt': 402, 'MMQ_ref': 60, 'MMQ_alt': 60, 'MPOS': 31, 'TLOD': 47.83000183105469, 'AD_ref': 15, 'AD_alt': 17, 'AF': 0.5289999842643738, 'DP': 32, 'FAD_ref': 15, 'FAD_alt': 17, 'AD': 1, 'SOR': -0.10018000635055137}\n",
            "\n",
            "\n",
            "\n",
            "{'chrom': 'chr16', 'pos': 11414455, 'ref': 'T', 'alts': ('C',), 'b64encode': 'MTZfMTE0MTQ0NTVfVF9D', 'filter': 'germline', 'any_pub': nan, 'is_molecularly_validated_pzm': nan, 'all_cohort_af': 0.4623270034790039, 'ECNT': 1, 'LCR': False, 'POPAF': 0.11100000143051147, 'RMCL': 21, 'SEGDUP': False, 'SIMPLEREP': False, 'STR': False, 'WESREG': True, 'is_snv': True}\n",
            "\n",
            "\n",
            "\n",
            "***************\n",
            "***************\n",
            "+++++++++++++\n",
            "------- appending\n",
            "{'chrom': 'chr1', 'pos': 116579678, 'ref': 'C', 'alts': ('CTCA',), 'b64encode': 'MV8xMTY1Nzk2NzhfQ19DVENB', 'filter': 'clustered_events;germline;haplotype', 'any_pub': nan, 'is_molecularly_validated_pzm': nan, 'GERMQ': 1, 'MBQ_ref': 30, 'MBQ_alt': 30, 'MFRL_ref': 410, 'MFRL_alt': 420, 'MMQ_ref': 60, 'MMQ_alt': 60, 'MPOS': 46, 'TLOD': 25.309999465942383, 'AD_ref': 25, 'AD_alt': 13, 'AF': 0.34599998593330383, 'DP': 38, 'FAD_ref': 24, 'FAD_alt': 13, 'AD': 1, 'SOR': -0.32482161940123755}\n",
            "\n",
            "\n",
            "\n",
            "{'chrom': 'chr1', 'pos': 116579678, 'ref': 'C', 'alts': ('CTCA',), 'b64encode': 'MV8xMTY1Nzk2NzhfQ19DVENB', 'filter': 'clustered_events;germline;haplotype', 'any_pub': nan, 'is_molecularly_validated_pzm': nan, 'all_cohort_af': 0.0007688360055908561, 'ECNT': 3, 'LCR': False, 'POPAF': 7.300000190734863, 'RMCL': 14, 'SEGDUP': False, 'SIMPLEREP': False, 'STR': False, 'WESREG': True, 'is_snv': True}\n",
            "\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# TEST\n",
        "if run_test:\n",
        "  _test_sample_metrics_ndf.head()"
      ],
      "metadata": {
        "id": "4Mre-02KL3MG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# TEST\n",
        "if run_test:\n",
        "  _test_sample_metrics_ndf.dtypes"
      ],
      "metadata": {
        "id": "sPWTSmTFUH-V"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# TEST\n",
        "if run_test:\n",
        "  _test_site_metrics_ndf.head()"
      ],
      "metadata": {
        "id": "lMOet_3yL516"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# TEST\n",
        "if run_test:\n",
        "  _test_site_metrics_ndf.dtypes"
      ],
      "metadata": {
        "id": "EO1CrGpLUR3a"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def to_csv_to_gs(df, filename):\n",
        "  filename = filename + \".csv.gz\"\n",
        "  full_filename = os.path.join(working_dir, filename)\n",
        "  # compression level for gzip ranges from 0 to 9 with 9 being highest compression rate and slowest.\n",
        "  df.to_csv(full_filename, index=False, compression={\"method\": \"gzip\", \"compresslevel\": 9})\n",
        "  # obj = f\"{gs_bucket.rstrip('//')}/dfs/{filename}\"\n",
        "  # copy_object(full_filename, obj)\n",
        "  # return obj\n",
        "  copy_object_to_bucket(\"broad-dsde-methods-vj\", f\"pzm/dfs/{filename}\", full_filename)\n"
      ],
      "metadata": {
        "id": "OcozRJYZddzx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# TEST\n",
        "if run_test:\n",
        "  to_csv_to_gs(_test_sample_metrics_ndf, _test_sample_id)\n",
        "  to_csv_to_gs(_test_site_metrics_ndf, _test_sample_id)"
      ],
      "metadata": {
        "id": "_sk6LKGzRga6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "counts_dfs = []\n",
        "for id in tqdm(sample_ids):\n",
        "  # print(f\"Processing {id} [{counter + 1}/{len(sample_ids)}]...\")\n",
        "  filename, unfiltered_filename = localize_obj_if_needed(id)\n",
        "  sample_metrics_df, sample_metrics_ndf, site_metrics_df, site_metrics_ndf = get_normalized_df(id, filename, unfiltered_filename)\n",
        "  counts_df = print_summary_stats(sample_metrics_df, site_metrics_df)\n",
        "\n",
        "  # counts_df[\"sample_id\"] = id\n",
        "  # counts_dfs.append(counts_df.copy())\n",
        "  # Return a copy of `counts_df` where a new column named `sample_id` with\n",
        "  # the value `id` in every row, is added to the dataframe. Then add the\n",
        "  # new dataframe to the list of dataframes.\n",
        "  counts_dfs.append(counts_df.assign(sample_id=id))\n",
        "\n",
        "  to_csv_to_gs(sample_metrics_df, id + \"_sample_metrics\")\n",
        "  to_csv_to_gs(sample_metrics_ndf, id + \"_normalized_sample_metrics\")\n",
        "\n",
        "  to_csv_to_gs(site_metrics_df, id + \"_site_metrics\")\n",
        "  to_csv_to_gs(site_metrics_ndf, id + \"_normalized_site_metrics\")\n",
        "\n",
        "  # os.remove(filename)\n",
        "  # os.remove(unfiltered_filename)\n",
        "  # print(f\"Finished processing {id}.\")\n",
        "  # print(\"-\" * 70 + \"\\n\")"
      ],
      "metadata": {
        "id": "4YSGhHRHXsGd"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "counts_df = pd.concat(counts_dfs, ignore_index=True)[[\"sample_id\", \"site_or_sample_level\", \"category\", \"value\", \"count\"]]\n",
        "obj = to_csv_to_gs(counts_df, \"counts\")\n",
        "print(f\"Pushed counts csv to: {obj}\")\n",
        "print(counts_df.shape)\n",
        "counts_df.head()"
      ],
      "metadata": {
        "id": "i3JtICzxvzax",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 241
        },
        "outputId": "72067492-99cb-4f23-a1ed-a5cd375cdee5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Pushed counts csv to: None\n",
            "(14416, 5)\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "   sample_id site_or_sample_level                      category     value  \\\n",
              "0  SS0012986               sample                       any_pub       NaN   \n",
              "1  SS0012986               sample                       any_pub    mosaic   \n",
              "2  SS0012986               sample                       any_pub  germline   \n",
              "3  SS0012986               sample                        filter      PASS   \n",
              "4  SS0012986               sample  is_molecularly_validated_pzm       NaN   \n",
              "\n",
              "   count  \n",
              "0  20229  \n",
              "1      1  \n",
              "2      1  \n",
              "3  20231  \n",
              "4  20229  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-2ded242a-9c44-4fca-81dd-bdbbe9297a90\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sample_id</th>\n",
              "      <th>site_or_sample_level</th>\n",
              "      <th>category</th>\n",
              "      <th>value</th>\n",
              "      <th>count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SS0012986</td>\n",
              "      <td>sample</td>\n",
              "      <td>any_pub</td>\n",
              "      <td>NaN</td>\n",
              "      <td>20229</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SS0012986</td>\n",
              "      <td>sample</td>\n",
              "      <td>any_pub</td>\n",
              "      <td>mosaic</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SS0012986</td>\n",
              "      <td>sample</td>\n",
              "      <td>any_pub</td>\n",
              "      <td>germline</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SS0012986</td>\n",
              "      <td>sample</td>\n",
              "      <td>filter</td>\n",
              "      <td>PASS</td>\n",
              "      <td>20231</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SS0012986</td>\n",
              "      <td>sample</td>\n",
              "      <td>is_molecularly_validated_pzm</td>\n",
              "      <td>NaN</td>\n",
              "      <td>20229</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-2ded242a-9c44-4fca-81dd-bdbbe9297a90')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-2ded242a-9c44-4fca-81dd-bdbbe9297a90 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-2ded242a-9c44-4fca-81dd-bdbbe9297a90');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 52
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# fig, axes = plt.subplots(ncols=2, nrows=1, figsize=(10, 5), sharey=True)\n",
        "# plt.subplots_adjust(wspace=0.05)\n",
        "# counts_df_copy = counts_df.copy()\n",
        "# counts_df_copy = counts_df_copy.fillna(\"Unknown\")\n",
        "\n",
        "# hue_order = [\"site\", \"sample\"]\n",
        "# palette = {\"site\": \"red\", \"sample\": \"blue\"}\n",
        "# colors = sns.color_palette(\"Paired\")\n",
        "# colors = [colors[2], colors[3]]\n",
        "\n",
        "# def plot(title, data, ax):\n",
        "#   # sns.boxplot(data=data, x=\"value\", y=\"count\", hue=\"site_or_sample_level\", ax=ax, dodge=0.4, palette=colors, hue_order=hue_order)\n",
        "#               #width=0.5, boxprops=dict(linewidth=1, facecolor='w', edgecolor='k', alpha=1),\n",
        "#                   #whiskerprops=dict(linewidth=0.75, color='k', alpha=1), capprops=dict(linewidth=0.5, color='k', alpha=1))\n",
        "#   # sns.swarmplot(data=data, x=\"value\", y=\"count\", hue=\"site_or_sample_level\", color=\".25\", ax=ax, dodge=0.2, hue_order=hue_order) #, palette=colors,)\n",
        "#   sns.violinplot(data=data, x=\"value\", y=\"count\", hue=\"site_or_sample_level\", ax=ax, dodge=0.2, hue_order=hue_order, palette=colors,)\n",
        "#   ax.legend().set_visible(False)\n",
        "#   ax.set(ylabel=None)\n",
        "#   ax.set_title(title)\n",
        "\n",
        "# plot(\"Molecularly validated variant\\ndistribution across samples\", counts_df_copy[counts_df_copy[\"category\"]==\"is_molecularly_validated_pzm\"], axes[0])\n",
        "# plot(\"Mutect2 filter distribution across samples\", counts_df_copy[counts_df_copy[\"category\"]==\"filter\"], axes[1])\n",
        "\n",
        "# # ax.set_yscale(\"log\")\n",
        "# axes[0].set_xlabel(\"\\nIs molecularly validated as PZM?\")\n",
        "# axes[0].set_ylabel(\"Count\\n\")\n",
        "# axes[1].set_xlabel(\"\\nMutect2 Filter\")\n",
        "\n",
        "# handles, labels = axes[0].get_legend_handles_labels()\n",
        "# legend_handles = [handles[0], handles[1]]\n",
        "# legend_labels = [\"Sample Metrics\", \"Site Metrics\"]\n",
        "# legend = plt.legend(legend_labels, bbox_to_anchor=(0.0, -0.35), loc=\"lower center\", ncol=2)\n",
        "\n",
        "# plt.show()\n"
      ],
      "metadata": {
        "id": "N8w6rpKZf2K0"
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}